DOC-1860: 6.4 Enterprise Docs release (#2655)
tinymce/tinymce-docs: modules/ROOT/pages/6.4.1-release-notes.adoc

As a result, in {productname} 6.4.1, visual characters no longer render inside `contenteditable="false"` elements.

=== Lists in a noneditable root were incorrectly editable using list API commands, toolbar buttons and menu items
//#TINY-9458

In previous versions of {productname}, a bug was found that affected the *lists* and *Advanced List* plugins when the {productname} editor body was set to `contenteditable = "false"`. The consequence meant that the lists, that were even set as `contenteditable="false"`, elements could be modified by using list API commands, toolbar buttons and menu items.

The fix involved implementing a new process to  check the element conditions. The check now looks to identify if the *list* or *Advanced List* is a `contenteditable = "false"` element.

As a result, in {productname} 6.4.1, *list* or *Advanced List*  plugins that are set as `contenteditable = "false"` elements can no longer be modified by using list API commands, toolbar buttons and menu items.

=== Color picker dialog would not update the preview color if the hex input value was prefixed with # symbol
//#TINY-9457

> In previous versions of {productname}, the preview sample box within the color picker dialog would not render. This issue could be replicated when a hex `value` was prefixed with a `#` symbol.
> 
> Example of a `incorrect` vs `correct` value input:
> 
> * `#BDE123`: Incorrect hex value input.
> * `BDE123`: Correct hex value input.
> 
> In order to fix this bug, {productname} 6.4.1 now detects if the user inserts a `#` symbol before the hex `value`, and removes the leading `#` prefix before rendering to the color picker component.
> 
> In {productname} 6.4.1, the color picker works as expected when using hex `values` with the `#` prefix.

Author: ```suggestion
In previous versions of {productname}, the preview sample box within the color picker dialog would not render. This issue could be replicated when a hex `value` was prefixed with a `+#+` symbol.

Example of a `incorrect` vs `correct` value input:

* `#BDE123`: Incorrect hex value input.
* `BDE123`: Correct hex value input.

In order to fix this bug, {productname} 6.4.1 now detects if the user inserts a `+#+` symbol before the hex `value`, and removes the leading `+#+` prefix before rendering to the color picker component.

In {productname} 6.4.1, the color picker works as expected when using hex `values` with the `+#+` prefix.
```

=== The "Edit Link" dialog incorrectly retrieved the URL value when opened immediately after the link insertion
//#TINY-7993

A bug was found within the context menu and dialog when inserting `links` in previous versions of {productname}. The issue resulted in the `open link` being deactivated within the context menu and dialog when the user would execute a specific chain of clicks. Such as:

.Example
====
. Opening the editor
. Type “test1 test2 test3” into the editor
. Double click to select “test2”
. Use the context menu to insert a new link over the text “test2”
. Add a valid URL
. Press `Enter`
. Without taking any other action, right click on “test2”

In this example the bug would deactivate the `open link` button within the context menu and dialog, which removes the users ability to reopen the “open link” dialog.
====

The fix now allows all three link-plugins (Link, Remove link and Open link) to work within the context menu items.

In {productname} 6.4.1, when the user executes right `mouse click` on the link selection immediately after attaching a link to a `<string>`, the link-plugins now work as expected.

> === Editor commands ForwardDelete and Delete were deleting contents inside noneditable elements
> //#TINY-9477
> 
> Previously, when using the `ForwardDelete` or `Delete` commands in earlier versions of {productname}, the contents within `contenteditable="false"` elements were deleted. This meant that if there were nested elements within a `contenteditable="false"` root element that were also set to `contenteditable="false"`, they could be deleted from the root element.
> 
> To resolve this issue, {productname} 6.4.1 included several enhancements to prevent all commands and keyboard functions, including delete/forwarddelete, from affecting `contenteditable="false"` content.
> 
> Consequently, in {productname} 6.4.1, using the forwarddelete or delete keys on `contenteditable="false"` elements within a `contenteditable="false"` element will no longer delete the contents within the `contenteditable="false"` elements.
> 
> === Backspace or delete keys was deleting contents inside noneditable elements
> //#TINY-9477
> 
> In the previous versions of {productname}, when utilizing the backspace or delete keys, they would delete the contents inside `contenteditable="false"` elements.
> 
> As a consequence, elements nested within a `contenteditable="false"` root element set to `contenteditable="false"`, were able to be removed/deleted from the `contenteditable="false"` root element. To fix this issue, {productname} 6.4.1 introduced various improvements to block all commands and keyboard functions such as `delete/backspace` from being applied to `contenteditable="false"` content. 
> 
> 
> As a result, in {productname} 6.4.1, when using the backspace or delete function keys on `contenteditable="false"` elements within a `contenteditable="false"` element, the contents within the `contenteditable="false"` elements will no longer be deleted. 
> 

Author: ```suggestion
=== Editor commands ForwardDelete, Backspace and Delete were deleting contents inside`contenteditable="false"` elements
//#TINY-9477

Previously, when using the ForwardDelete, Backspace or Delete commands in earlier versions of {productname}, the contents of a `contenteditable="false"` nested element would be deleted.

As a consequence, if a `contenteditable="false"` element was nested within a `contenteditable="false"` root element, the contents of the nested element would be deleted from the root element when the specified commands were executed.

To resolve this issue, {productname} 6.4.1 included several enhancements to prevent all commands and keyboard functions which include ForwardDelete, Backspace and Delete commands from affecting `contenteditable="false"` content.

```

Author: @TheSpyder Please see above suggested merge for the above two sections.

Reviewer: Almost. `backspace` and `delete` are keys on the keyboard, `Delete` and `ForwardDelete` are commands in the editor. You've made Backspace sound like a command and incorrectly combined the two delete concepts.

Author: @TheSpyder Please see above edited changes.

Reviewer: Minor corrections and fixing a typo

```
=== Content could be deleted inside `contenteditable="false"` elements
//#TINY-9477
When content was deleted (either through the ForwardDelete/Delete commands, or the user pressing Backspace/Delete keys) in earlier versions of {productname}, the contents of a `contenteditable="false"` nested element would be deleted.
As a consequence, if a `contenteditable="false"` element was nested within a `contenteditable="false"` root element, the contents of the nested element would be deleted from the root element when the specified command or keys were executed.
To resolve this issue, {productname} 6.4.1 includes several enhancements to prevent all commands and keyboard functions from affecting `contenteditable="false"` content which includes the ForwardDelete/Delete commands and both the Backspace and Delete keys.
```

To fix this issue, {productname} 6.4.1 fixed this issue by addressing the conflict between the document skin's CSS and the autoresize bottom margin setting. After this fix, the resize function now checks for loops and considers the margin bottom as always set to 0 instead of autoresize bottom margin, which ensutes that the editor's height is calculated correctly. 

=== Plugin - Search and Replace - Throws a JS console error if you search for a string that doesn't exist and try to click on the preferences button.
//#TINY-9686

In previous versions of {productname}, the 'Search and Replace' plugin had a known issue that caused a console error when searching for a string that wasn't in the editor content, and then attempting to click on the `preferences` button.

As a result, an "incorrect type" error would prevent users from opening the preferences section.

In {productname} 6.4.1, this issue has been resolved.

=== Toolbar doesn't hide when editor is out of view in scrollable container
//#TINY-9652

In previous versions of {productname}, when the editor is within a scrollable container, the toolbar would remain visible when the editor is scrolled out of view.

> With the release of {productname} 6.4.1, the new `ui_mode` option provides improvements to editors within scrollable contexts, including hiding the toolbar when the editor is out of view when the `ui_mode` option is set to 'split'.
> 
> === `color_cols`: 0 caused the page to crash
> //#TINY-9649
> 
> In previous versions of {productname}, the color swatch would attempt to display 0 columns of colors when the relevant `color_cols`, `color_cols_foreground` or `color_cols_background` options were set to 0. As a consequence, the tab would crash when opening that color swatch.

Reviewer: It was fixed in 6.3, so I figured we should remove the 6.4.1 mention...

```suggestion
=== Editor increases the height indefinitely when using autoresize with `content_css: 'document'`
//#TINY-8872

In previous versions of {productname}, it was discovered that the editor had some issues with the autoresize plugin when using `content_css: 'document'`, which caused the editor height to expand indefinitely.

To fix this issue, {productname} addressed the conflict between the document skin's CSS and the autoresize bottom margin setting. After this fix, the resize function now checks for loops and considers the margin bottom as always set to 0 instead of autoresize bottom margin, which ensutes that the editor's height is calculated correctly. 
```

//#TINY-9449

In previous versions of {productname}, contenteditable="false" elements such as a *merge tags* could clone the styling to a new table row. In some situations, this cloning would happen during the creation of new table row, where the merge tag prefix would unexpectedly clone the formatting from the previous cell to the new cell.

To correct this issue, {productname} added changes to the *clone logic*, to not copy the format for `contenteditable="false"` elements.

In {productname} 6.4.1, the formatting of `contenteditable="false"` elements are no longer cloned to new cells while creating new table rows.

=== Changed the color of @dialog-table-border-color, and added right padding to the first cell of dialog table
//#TINY-9380

An area for improvement was identified during an accessibility audit for the *help menu* within {productname}, which focuses on users that rely on **zoom orient** to help with visibility.

The consequence of the previous UX, meant that adjustments needed to be made to the horizontal lines within the *help menu* table, that seperate each `row`.

> 

Reviewer: ```suggestion
In previous versions of {productname}, contenteditable="false" elements such as a *merge tag* could clone the styling to a new table row. In some situations, this cloning would happen during the creation of new table row, where the merge tag prefix would unexpectedly clone the formatting from the previous cell to the new cell.
```

For information on the **Checklist** plugin, see: xref:checklist.adoc[Checklist].

=== Merge Tags 1.2.1

The {productname} 6.4.1 release includes an accompanying release of the **Merge Tags** premium plugin.

**Merge Tags** 1.2.1 includes the following change.

==== The mceMergeTag command now preserves the selected content as a plain text and wraps it in a merge tag prefix and suffix if the content length exceeds the limit.
//#TINY-9444

In previous versions of {productname}, the prefix and suffix were removed when merge tag content exceeded the character limit for the merge tag.

In {productname} 6.4.1 this content is retained, as well as the prefix and suffix, though it is not converted into a Merge Tag.

[NOTE]
Any content exceeding the character limit for Merge Tags will not be converted into a Merge Tag.

For information on the **Merge Tags** plugin, see: xref:mergetags.adoc[Merge Tags].

=== PowerPaste 6.1.4

The {productname} 6.4.1 release includes an accompanying release of the **PowerPaste** premium plugin.

**PowerPaste** 6.1.4 includes the following fixes.

==== White-space style property stripped from copied Excel content by word import
//#TINY-9607

> The white-space style property, which determines how white space characters are handled in HTML, was causing issues when content was copied from Excel and pasted into the {productname} editor. The problem occurred in Chrome and Firefox, but was not visable in Safari. Upon import, the white-space style property was being stripped from the content, resulting in a loss of formatting.

Reviewer: ```suggestion
The white-space style property, which determines how white space characters are handled in HTML including word wrap, was causing issues when content with word wrap set was copied from Excel and pasted into the {productname} editor. The problem occurred in Chrome and Firefox, but was not visible in Safari. Upon import, the `white-space` style property was being stripped from the content, resulting in a loss of word wrap.
```

. Reading unsanitized HTML when previewing templates and when `editor.dom.create` is run during insertion may allow room for cross-site scripting (XSS), which can be a security issue.

To address this, the Template plugin in {productname} 6.4.1 now parses and serializes template HTMLs before it is previewed and before `editor.dom.create` is run on it during template insertion. By doing this, all template previews will now visually match the post-insertion result, and the possibility of XSS issues arising is minimized.

=== Pressing backspace in an empty line now preserves formatting in a previous empty line
//#TINY-9454

In previous versions of {productname}, when pressing backspace in an empty line with another empty line containing inline formats above it, the deletion handler was found to be first removing both lines and then inserting a newline. As a consequence, the existing inline formats of the previous empty line was not preserved.

In {productname} 6.4.1, pressing backspace in an empty line now preserves the inline formats of the previous empty line.

> === Pressing enter inside the `fontsizeinput` input would not move the focus back into the editor content
> //#TINY-9598
> 
> During development for the new `fontsizeinput` toolbar item, an issue was found that by pressing enter inside the input field would not move the `focus` back into the {productname} editor body. 
> 
> As a consequence, any changes made to the font size within the new toolbar input field, would result in the user having no way to use the new font size within the editor.
> 
> After some improvements to the `fontsizeinput` toolbar item, pressing enter in the input field now sets the focus back to the {productname} editor viewport, which allows the new font size to be used. 
> 
> In {productname} 6.4.1, pressing enter inside the `fontsizeinput` input now returns the focus back to the editor content body.
> 
> 

Reviewer: This feature is new in 6.4

```suggestion
```

.Setting `highlight_on_focus: true` option to tinymce.init().
[source, js]
----
tinymce.init({
    selector: "textarea",
    highlight_on_focus: true
});
----

=== New fontsizeinput toolbar item which allows the user to set the size via input and also increase and decrease it with + and - buttons
//#TINY-9429

In previous versions of {productname}, users did not have the ability to set their font size by either direct input into a component field or by using plus `+` or minus `-` buttons.

As a consequence, {productname} introduced a new UI component that gives users an option to set the font sizes via the `fontsizeinput` toolbar button. For example:

[source,js]
----
tinymce.init({
    selector: "textarea",  // change this value according to your HTML
    toolbar: "fontsizeinput",
    font_size_input_default_unit: "pt"
});
----

[NOTE]
More details about this component and the possible values that can be used within the new `fontsizeinput` input field can be found in `xref:available-toolbar-buttons.adoc#the-core-toolbar-buttons[fontsizeinput]`.

> === Added skipFocus option to the ToggleToolbarDrawer command to preserve focus

Reviewer: I always forget to check the heading which is _above_ the jira tag comment...

```suggestion
=== Added `skip_focus` check to the ToggleToolbarDrawer command
```

//#TINY-9337

> A focusing event that was in the initial *FloatingToolbarButton* config, in previous versions of {productname} in (2019) was causing `focus` issues with the *ToggleToolbarDrawer*. As a consequence, this meant that while the editor was scrolled into `focus`(viewport position), and the *ToggleToolbarDrawer* command `skipFocus: boolean` was set, the toolbar draw could still take `focus` when it was not meant to.
> 
> In {productname} 6.4.1, an improvement was made to the focusing `event` which is now removed after toggling the toolbar.
> 
> As a result to the fix, while the *FloatingToolbarButton* `focus` is initiated, the toolbar now focuses correctly without stealing `focus` when `skipFocus: 'boolean' is set.

Reviewer: That was a decent job of describing the code change, but those names do not exist on our API and have no meaning to customers. Here's what changed from an integrator perspective:

```suggestion
In previous versions of {productname}, the *ToggleToolbarDrawer* command contained an incorrect focus call. As a consequence, this meant that when the *ToggleToolbarDrawer* command option `skip_focus: true` was set the toolbar drawer would take `focus` when it was not meant to.

In {productname} 6.4.1, an improvement was made to the focusing code which now checks for `skip_focus`.

As a result of the fix, when using the *ToggleToolbarDrawer* command, the toolbar now avoids stealing focus when `skip_focus: true` is set.
```

=== The sidebar element now has accessibility role region when visible and presentation when hidden
//#TINY-9517

In previous versions of {productname}, Axe-DevTools (web accessibility testing extension) identified that the sidebar element within {productname} was set to `role="complementary"` even when the element was hidden. This created issues with screen readers, as the general rule is that landmarks should not be contained within another landmark.

As a consequence by having the sidebar set to `role="complementary"`, tools such as screen readers in some situations would disable their functionality.

To fix this issue, changes were made to the role attribute of the sidebar element, which is now set to `role="region"` when it's active state is visible, otherwise it defaults to `role="presentation"` when hidden.

This is important as the role of `region` ensures the sidebar and its contents are accessible to screen readers and other accessibility tools. When the sidebar element is hidden, it has a role of `presentation`, which is used to inform accessibility tools that the element can be safely ignored. 

> In {productname} 6.4.1, when the sidebar is either visible or hidden it now passes the accessibility test with Axe-DevTools.

Author: ```suggestion
In {productname} 6.4.1, when the sidebar is either visible or hidden it now passes all the identified accessibility requirements.
```

=== Updated the Code Sample and Template plugin dropdown menu styles to match the default {productname} styles
//#TINY-9630

Previously, the Code Sample and Template open source plugins used browser-native elements for some dropdown menus.

With this update, these menus are now created using the equivalent {productname} custom element, making these dropdown menus present with the same styling as other {productname} UI elements.

[[bug-fixes]]
== Bug fixes

{productname} 6.4.1 also includes the following bug fixes:

> === Sometimes the editor would finish initializing before the theme would have finished loading
> //#TINY-9556
> 
> In earlier versions of {productname}, the properties set for the `renderUI` could cause conflicts with the theme load times. As a consequence, the `renderUI` would finish rendering before the {productname} theme it relied upon was initialized.
> 
> To fix this bug, {productname}, made changes to the `renderUI` properties that allowed the theme to return a Promise, resolving the load ordering issues. 
> 
> In {productname} 6.4.1, all `renderUI` property loading issues have been resolved.
> 

Reviewer: I missed this when I corrected the other copy of it under additions (search for TINY-9556). We don't need two.

```suggestion
```

In {productname} 6.4.1, the color in the relevant toolbar button is updated when changing either the `forecolor` or `backcolor` via a menu item.

=== The onSetup api function would not run when defining custom group toolbar button
//#TINY-9496

A bug was discovered that affected the custom group toolbar button in {productname}, due to missing some event handlers. As a consequence, the `onSetup` api function would not run as intended when defining a custom group toolbar button.

In {productname} 6.4.1, this bug was fixed by adding a handler to the toolbar group button that now calls the `api` as intended.

As a result, the `onSetup` function for custom group toolbar buttons now runs correctly in {productname} 6.4.1.

=== An element could be dropped onto the descendants of an element with a `contenteditable="false"` attribute
//#TINY-9364

In {productname} 6.0, the drag and drop `config` did not check some conditions surrounding element placements using either a drag or drop action.  This affected users running {productname} 6.0, which would effectively allow descendants of `contenteditable="false"` element to be dropped into another `contenteditable="false"` element.

> Fixes were made to ensure that `dragging | dropping` onto `contenteditable="false"` element targets on all browsers is `disabled`. This also includes when the drop target cursor is inside the `contenteditable="true"` element, or just before it.

Reviewer: `dragging | dropping` doesn't seem appropriate here? just plain english "dragging or dropping"? something like that?

Author: ```suggestion
Fixes were made to ensure that `dragging or dropping` onto `contenteditable="false"` element targets on all browsers is `disabled`. This also includes when the drop target cursor is inside the `contenteditable="true"` element, or just before it.
```

The fix involved implementing a new process to  check the element conditions. The check now looks to identify if the *list* or *Advanced List* is a `contenteditable = "false"` element.

As a result, in {productname} 6.4.1, *list* or *Advanced List*  plugins that are set as `contenteditable = "false"` elements can no longer be modified by using list API commands, toolbar buttons and menu items.

=== Color picker dialog would not update the preview color if the hex input value was prefixed with # symbol
//#TINY-9457

In previous versions of {productname}, the preview sample box within the color picker dialog would not render. This issue could be replicated when a hex `value` was prefixed with a `#` symbol.

Example of a `incorrect` vs `correct` value input:

* `#BDE123`: Incorrect hex value input.
* `BDE123`: Correct hex value input.

> In order to fix this bug, {productname} 6.4.1, not detects if the user inserts a `#` symbol before the hex `value`, which removes the leading `#` prefix before rendering to the color picker component.

Reviewer: ```suggestion
In order to fix this bug, {productname} 6.4.1 now detects if the user inserts a `#` symbol before the hex `value`, and removes the leading `#` prefix before rendering to the color picker component.
```

In {productname} 6.4.1, when the user executes right `mouse click` on the link selection immediately after attaching a link to a `<string>`, the link-plugins now work as expected.

=== Editor commands ForwardDelete and Delete were deleting contents inside noneditable elements
//#TINY-9477

Previously, when using the `ForwardDelete` or `Delete` commands in earlier versions of {productname}, the contents within `contenteditable="false"` elements were deleted. This meant that if there were nested elements within a `contenteditable="false"` root element that were also set to `contenteditable="false"`, they could be deleted from the root element.

To resolve this issue, {productname} 6.4.1 included several enhancements to prevent all commands and keyboard functions, including delete/forwarddelete, from affecting `contenteditable="false"` content.

Consequently, in {productname} 6.4.1, using the forwarddelete or delete keys on `contenteditable="false"` elements within a `contenteditable="false"` element will no longer delete the contents within the `contenteditable="false"` elements.

> === Backspace or delete keys was deleting contents inside noneditable elements

Reviewer: I really don't think this needed two sections, but I'm not in a position to combine them myself today

{productname} 6.4.1 introduced a new behavior if an element is dropped onto a table border, it will now be inserted into the left or right adjacent cell instead of being deleted.

=== Noneditable elements would be removed when dragged and dropped within a noneditable root
//#TINY-9558

In previous version of {productname}, dragging ``contenteditable="false"` elements into a `contenteditable="false"` root would remove the elements `contenteditable="false"` attributes.

As this is not the expected behavior for a `contenteditable="false"` element or root, this bug in particular, created issues in situations such as: *(a)* when the editor body was set to `contenteditable="false"`, and *(b)* when a `contenteditable="false"` element is dragged and dropped into a table element that is within a `contenteditable="false"` root, which would result in the `contenteditable="false"` attributes blocks being deleted from the element that was dropped.

> To fix this issue, improvements were made to the drag and drop function that checks if the `dom` element `isEditable` to its `parentElement`.

Reviewer: ```suggestion
To fix this issue, improvements were made to the drag and drop function that checks if the source and target elements are within editable sections. This ensures the correct logic runs during a drop event and avoids the attribute removal.
```

In previous versions of {productname}, the annotations behavior was not working as expected when using `comments`.

As a consequence, the annotation (highlighted-text-selection) would not be removed from the text selection, if the annotation was immediately deleted after being created.

While applying a fix for this issue, it was identified that this was a regression created when {productname} fixed issue: _`Change the annotation api to avoid dom mutation in plugins using it`_, which adjusts the selection in the way that it is _impossible_, to retain the selected word. 

However, with the release of {productname} 6.4.1, after immediately deleting annotations after they are created, they will no longer will be retained on the text selection.

=== Inserting a link for a selection from quickbars didn't preserve formatting
//#TINY-9593

In previous versions of {productname}, inserting a link for a text selection from `quickbars` did not preserve formatting.

As a consequence, when formatting the selection such as with `*bold*` or `_italic_`, then inserting the link via the `quickbars` plugin, the link would result in the applied formatting been removed from the text selection.

> To fix this issue, improvements were made to the conditional check that is responsible for returning `optionals` when the anchor is called to gather the text selection.

Reviewer: This item was close, but the last line went a bit deep into the weeds.

```suggestion
To fix this issue, improvements were made to a conditional check that is responsible for locating anchors in text selection.
```

As a result, formatting after pasting a link from `quickbars` is now preserved to the selection {productname} 6.4.1.

=== Inline dialog position was not correct when the editor wasn't inline and was contained in a fixed or absolute positioned element
//#TINY-9554

> In previous versions of {productname}, a bug was found that affected the inline dialog position for elements set to either a `fixed` or `absolute` position.  As a consequence, this bug in particular, presented itself when the editor wasn't `inline`, which resulted in the dialog being rendered outside of the {productname} viewport `bounds`.
> 
> A fix for this bugs was implemented in {productname} 6.4.1, that corrected this issue, resulting in the `inline` dialog working as expected with this release.

Reviewer 2: ```suggestion
In previous versions of {productname}, inline dialogs, such as the Search and Replace and Accessibility Checker dialogs, would be positioned incorrectly to overlap with the toolbar when the editor was in classic `iframe` mode and was contained in an parent element with a `fixed` or `absolute` position. Inline dialogs would be rendered out of the bounds of the editor's content area to overlap with the toolbar, causing a bug.

In {productname} 6.4.1, the container of the editor is used as the positioning bounds of inline dialogs when the editor is in classic mode. This ensures that the inline dialogs are positioned correctly when a classic-mode editor is in a container with a `fixed` or `absolute` position.
```

[source, js]
----
tinymce.init({
    selector: "textarea",
    plugins: "table code",
    toolbar: "forecolor backcolor tableprops",
    color_cols: 7, // <-- option was not respected
});
----

In {productname} 6.4.1, the color swatches opened by the `forecolor` and `backcolor` toolbar buttons have the amount of columns set by the `color_cols` option, if it is set.

=== Drag and dropping the last noneditable element out of its parent block would not properly pad the parent block element
//#TINY-9606

In previous versions of {productname}, dragging the last child element with a `contentedtiable="false"` attribute out of a block element would prevent the user from placing the caret back into that source block.

As a fix, {productname} 6.4.1, added a new condition which validates if the source element is empty when the drop operation occurs. If it is empty then it will be padded with a `BR` element so that it's possible to place the caret inside the block. 

In {productname} 6.4.1, dragging and dropping elements that are `contenteditable="false"` out of they're parent blocks now correctly pad the parent block element.

> === Nonbreaking space do not get highlighted when enabling show invisible characters
> //#TINY-9685
> 
> In previous versions of {productname}, it was discovered that nonbreaking spaces were not highlighting correctly when the "show invisible characters" plugin was enabled.
> 
> This issue caused not only confusion with the intended text layout, but impacted the overall user experience. To resolve this issue, {productname} introduced a fix to ensure that all invisible characters, including nonbreaking spaces, are accurately displayed when the plugin is enabled.

Reviewer: This is a confusing one to explain. Might need some rewording, there's too many negatives.

```suggestion
=== Non breaking spaces were not highlighted within a nested editable section
//#TINY-9685

In previous versions of {productname}, when using the "show invisible characters" feature of the `visualchars` plugin, non breaking spaces were not highlighting within an editable section that was nested inside a noneditable section.

To resolve this issue, {productname} 6.4.1 introduced a fix to ensure that all invisible characters, including nonbreaking spaces, are accurately displayed when the plugin is enabled.
```

Author: @TheSpyder Thanks for the above suggestion, It might be worth nothing that we are also trying to avoid using the term `noneditable` and `editable`, and replace with `contenteditable="false" or `contenteditable="true" from now on in the docs to be more consistent.

=== Dialog getting cut offscreen when opened from toolbar button in Inline Mode
//#TINY-9588

In previous versions of {productname}, inline dialogs would always open to the South West of the button used to open it from the toolbar, or South East with RTL (Right-to-Left) directionality. However, an issue was found where, when the toolbar was shorter than the inline dialog, inline dialogs would appear partially off screen when opened.

To address this issue, inline dialogs now determine a more appropriate direction to open, that is, any direction which will not cause the dialog to appear offscreen.

=== Checkmark icons are added to every color in color swatch
//#TINY-9567

In previous versions of {productname}, when using certain configurations, the color swatch feature could add a checkmark icon to every color, when the checkmark should only be displayed on a color when one is enabled.

In {productname} 6.4.1, the selector for applying the checkmark is updated so the checkmark will only be added and displayed for enabled colors.

=== The colour picker reset to default when formatting pre element with addtional format options
//#TINY-9342

> In previous versions of {productname}, formatting elements with a `pre-formatted` tag, and applying a background color from the color swatch would cause formatting issues when additional formatting options such as bold and italics were applied to the `pre-formatted` element. When these formatting conditions were combined, the color picker would reset the users defined color value, which would triggered the color picker to reset this value to default "000000".

Reviewer: ```suggestion
In previous versions of {productname}, formatting elements with a `pre-formatted` tag, and applying a background color from the color swatch would cause formatting issues when additional formatting options such as bold and italics were applied to the `pre-formatted` element. When these formatting conditions were combined, the color picker would reset the users defined color value, which would trigger the color picker to reset this value to default "000000".
```

To fix this {productname} 6.4.1, added a blocker for rendering fake carets for `contenteditable="false"` and table elements that are contained within other `contenteditable="false"` elements.

As a result, any `contenteditable="false"` or table elements that are nested within another `contenteditable="false"` element root will no longer have fake carets in {productname} 6.4.1.

=== Textareas with scrollbars in dialogs would not render rounded corners correctly on some browsers
//#TINY-9331

In previous versions of {productname}, textareas with scrollbars in dialogs would not render rounded corners correctly on some browsers, as the rounded corner would be overlaid by the right-angled corner of the scrollbar track. An example of a dialog affected by this issue was the source code dialog.

{productname} 6.4.1 added to the form textarea element a wrapper class containing border properties and an `overflow: hidden` property. This correctly clips the scrollbars to the border-radius on all browsers, resolving the bug.

> As a result, the `focus` dialog border now renders correctly in {productname} 6.4.1.

Reviewer 2: ```suggestion
```

This issue caused not only confusion with the intended text layout, but impacted the overall user experience. To resolve this issue, {productname} introduced a fix to ensure that all invisible characters, including nonbreaking spaces, are accurately displayed when the plugin is enabled.

=== Non breaking space is added after backspace below text pattern heading
//#TINY-9603

A recently discovered issue was identified in previous versions of {productname}, that affected the text formatting feature for **Text Pattern**. When a user would apply a text pattern heading such as `# Title`, followed by pressing "enter", the issue would the occur when the backspace key was pressed. As a result, a non-breaking space was being inserted to the beginning of the line instead. This unintended addition of a Non-breaking space caused formatting inconsistencies and errors.

In {productname} 6.4.1, pressing backspace after adding a Text Pattern, now formats the content as expected.

=== Incorrect event target for dragend and drop events when dragging and dropping noneditable elements within editor
//#TINY-9599

> A bug was discovered in previous version of {productname}, that would arise when dragging and dropping `contenteditable="false"` elements within the editor was that the target of the dragstart, dragend or drop events would be the editor instance itself, rather than the expected DOM target.
> 
> In {productname} 6.4.1, the events are now simulated based on the mouse event that produced them, meaning that the target and other properties, such as mouse coordinates, are now properly produced for the simulated drag and drop events.

Reviewer 2: ```suggestion
In previous versions of {productname}, when dragging and dropping `contenteditable="false"` elements within the editor, the target of the `dragstart`, `dragend`, and `drop` events would be the editor instance itself, rather than the expected DOM target.

In {productname} 6.4.1, the events are now simulated based on the mouse event that produced them, meaning that the target and other properties, such as mouse coordinates, are now properly produced for the simulated `dragstart`, `dragend`, and `drop` events.
```

NOTE: Although there was a Community release of {productname} 6.4, it was not released as a {enterpriseversion} or {cloudname} release. The first commercially supported release of {productname} post version 6.3.2 is {productname} 6.4.1.

[[new-premium-plugin]]
== New Premium Plugin

=== Advanced Template 1.1.0

The {productname} 6.4.1 release includes the release of a new premium plugin: **Advanced Template**.

The **Advanced Template** Premium plugin enables the creation of complex, interactive templates using {productname}.

The end-users can select content in a {productname} document and save the selection as a template via the {productname} user-interface.

For information on the **Advanced Template** plugin, see: xref:advanced-templates.adoc[Advanced Template].

[[accompanying-premium-plugin-changes]]
== Accompanying Premium Plugin changes

The following premium plugin updates were released alongside {productname} 6.4.1.

=== Advanced Code Editor 3.2.1

The {productname} 6.4.1 release includes an accompanying release of the **Advanced Code Editor** premium plugin.

**Advanced Code Editor** 3.2.1 includes the following improvements and additions.

> ==== New button to copy the current dialog content.

Reviewer 3: ```suggestion
==== New button to copy the current **Advanced Code Editor** content.
```

> This update adds a new **Copy code** button to the **Advanced Code Editor** for editors. This new **Copy code** button allows users to copy the entire contents of the dialog.

Reviewer 3: ```suggestion
This update adds a new **Copy code** button to the **Advanced Code Editor** for editors. This new **Copy code** button allows users to copy the entire contents of the **Advanced Code Editor**.
```

> ==== New buttons to increase and decrease the font size in the dialog.

Reviewer 3: ```suggestion
==== New buttons to increase and decrease the font size in the **Advanced Code Editor**.
```

> This update adds two new buttons for increasing and decreasing the font size of code displayed in the dialog.

Reviewer 3: ```suggestion
This update adds two new buttons for increasing and decreasing the font size of code displayed in the **Advanced Code Editor**.
```

==== New button to toggle dark mode.

> This update adds a new **Dark/light mode** button to toggle between either dark or light display modes for the dialog.

Reviewer 3: ```suggestion
This update adds a new **Dark/light mode** button to toggle between either dark or light display modes for the **Advanced Code Editor**.
```

//#TINY-9528

{productname} recently added a new *"fullscreen"* toggle button to the Advanced Code Editor Premium plugin, that gives users the ability to expand the Advanced Code Editor dialog toolbar in order to make better use of the available space on their screen, while providing them with a more comfortable working environment.

The toggle button is easily accessible for users that have the Advanced Code Editor Premium plugin by integrating it into their editor by using the `plugins: [ "fullscreen" ]` in pair with the below options.

.Example of fullscreen mode integration within {productname}.
[source, js]
----
tinymce.init({
  selector: "textarea",
  advcode_inline: true,
  plugins: [
    "code", "fullscreen", 'advcode',
  ],
  toolbar: "code"
});
----

As a result, the toggle button can be found in the top left corner of the Advanced Code Editor Premium toolbar. Once clicked, it will expand the dialog toolbar to fill the entire window, providing a more comprehensive view of the available options. A second click of the toggle button will return the dialog toolbar to its original size.

This option is now available in {productname} 6.4.1, as part of the core/ui API.

> === New text-size-increase and text-size-decrease icons
> //#TINY-9530
> 
> The {productname} package files have been updated with two new icons that enable users to increase or decrease the size of text in the Advanced Code Editor. The new icons are the `text-size-increase` and `text-size-decrease` icons. These new icons have been added to make it easier for users to adjust the size of the text within the code editor while making it easier to read and understand the code they are writing. 
> 
> .Example of `textSizeIncrease` and `textSizeDecrease` integration within {productname}.
> [source, js]
> ----
> tinymce.init({
>   selector: "textarea",
>   advcode_inline: true,
>   plugins: [
>     "code", "advcode", 
>   ],
>   toolbar: "textSizeIncrease textSizeDecrease"
> });
> ----
> [NOTE]
> This plugin feature is only available for paid {productname} subscriptions.

Reviewer 4: ```suggestion
```
Don't need to document new icons, and they're only included in the advanced code dialog, not available as toolbar buttons.

In previous versions of {productname}, the top right close 'x' button within the modal dialogs were not keyboard-navigable by the Tab key.

With the latest release of {productname} 6.4.1, users can now navigate through all dialogs that have a `x` close button. With these updates, users can now use the `Tab` key to move through the elements within the modal dialogue, making it more accessible and user-friendly. 

This change is applied to all dialogs that have a `x` close button within {productname}.

* Search and Replace
* Accessibility Checker
* Table Properties
* Help
* Source Code
* URL
* Image
* Url Modal Dialog toolbar buttons. 

[NOTE]
Its worth nothing that the expected result for the Alert and Confirm dialogs is that the `initial focus` is set on the most appropriate button when the dialog is opened; for example, the “OK” button for Alert dialogs, and the “Yes” button for Confirm dialogs.

In {productname} 6.4.1, these new additions/improvements are now available.

=== New ui_mode option for editor in scrollable containers support
//#TINY-9414

A new `ui_mode: 'split'` option for scrollable containers has been introduced with the release of {productname} 6.4.1.

> This new option is designed to give users a better UX, by allowing them to navigate content within their editor easier with the use of the new ui_mode when integrated into their editors. This can be achieved by adding the `ui_mode: 'split'` option to the `tinymce.init()` within the `<script>` tags as below:

Reviewer 4: ```suggestion
This new option configures the UI of the editor to allow scrolling popup menus. This prevents unexpected non-scrolling behaviour when within a scrolling context. When the editor is within a scrollable element add the `ui_mode: 'split'` option to the `tinymce.init()` within the `<script>` tags as below:
```

.Basic integration for new ui_mode option using `split`
[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your HTML
  ui_mode: 'split',
  min_height: 1400,
});
----

[TIP]

> Its worth noting that when setting the new `ui_mode: 'split'` option for scrollable containers, when setting the height for the `<div class="scrollable-container">` ensure that the height is set to be less than `<` the `min_height` set of the {productname} editor container and by applying the `overflow: scroll;` CSS bevaviour to the `<div class="scrollable-container">`.

Reviewer 4: ```suggestion
Its worth noting that when setting the new `ui_mode: 'split'` option for editors within scrollable containers, such as within elements with the `overflow: scroll;` CSS property, ensure that the `min_height` of the editor is greater than the height of the outer container.
```
@shanmen-tiny What would this be for specifically, when the editor is in a scrolling container?

Reviewer 5: @MitchCTiny @kemister85 I was talking to Karl before, and the content in the scrollable container needs to be greater than the height of the scrollable container so that the changes for `ui_mode: 'split'` is visible, and by setting the `min_height` of the editor.

Reviewer 4: ```suggestion
The new `ui_mode: 'split'` option alters the behaviour of editors within scrollable containers, where the `min_height` of the editor is greater than the height of the container element which has the `overflow="scroll"` style attribute.
```
Does this make more sense then?

Author: Im happy with that change. @shanmen-tiny happy with this?

The docking behavior of UI elements is usually only triggered when the page is scrolled. This means that the UI element will stay in the same place relative to the page, no matter how much you scroll. However, with the introduction of the 'split' UI mode, it is now possible to dock the UI element on both page and container scroll. This means that the UI element will stay in the same place relative to the container, regardless of how much you scroll the page or the container itself.

[NOTE]
* If you're using classic mode and sticky toolbars, the toolbar will dock, but it won't be entirely visible in either the window or the scrollable element. This applies to both viewports.
* In inline mode, the toolbar will also dock, but it won't be entirely visible in either the window or the scrollable element.
* ShadowDom is partially supported in this case, where the scrollable container is inside the shadow Root. But in cases when the scrollable container is outside the shadow Root (or an ancestor of it) it is not currently supported within versions of {productname} 6.4.1 or prior.

[[changes]]
== Changes

{productname} 6.4.1 also incorporates the following changes:

> === The link plugins context menu items will no longer appear for noneditable links

Reviewer 2: ```suggestion
=== The Link plugin's context menu items will no longer appear for `contenteditable="false"` links
```

//#TINY-9491

> In previous versions of {productname}, the link plugin would display the context menu items for `contenteditable="false"` links.
> 
> As a consequence, links such as `table of contents` would open the context menu when `clicked`. The behavior for the `table of contents`, is not designed to allow links to be modified or removed as the `table of contents` plugin is considered as a `contenteditable="false"` element.
> 
> To fix this bug, {productname} 6.4.1 introduced a new check that now verify if the link plugin is opened, which will now block the context menu on `table of contents` from opening upon when the `toc` is clicked. 
> 
> As a result, the context menu for the `table of contents` now behaves as expected in {productname} 6.4.1.

Reviewer 2: ```suggestion
In previous versions of {productname}, the context menu for `contenteditable="false"` anchor links would display items from the Link plugin, including "Link" and "Remove Link". This implies that editing or removing `contenteditable="false"` links are possible, which is not true. An example of an affected `contenteditable="false"` link is the Table of Contents links.

To resolve this undesired behavior, {productname} 6.4.1 introduced a new check that blocks the Link plugin from displaying its context menu items if the selection is on a `contenteditable="false"` anchor link.
```

As a consequence by having the sidebar set to `role="complementary"` for both states, tools such as screen readers in some situations would disable their functionality.

This is important as when a sidebar element is visible, it has an accessibility role of `region`. This ensures the sidebar and its contents are accessible to screen readers and other accessibility tools. This role is important for screen readers to identify the sidebar as a meaningful part of the page. When the sidebar element is hidden, it has a role of presentation, which is used to inform the user that the element is not important and can be safely ignored. 

To fix this issue, changes were made to the role attribute of the sidebar element, which is now set to `role="region"` when it's active state is visible, otherwise it defaults to `role="presentation"` when hidden.

> In {productname} 6.4.1, when the sidebar is either visible or hidden it now passes the accessibility test with Axe-DevTools.

Author: @MitchCTiny this was referenced from section 4. Result in the documentation in TINY-9517 as verified. @shanmen-tiny can you confirm?

> In {productname} 6.4.1, when the sidebar is either visible or hidden it now passes the accessibility test with Axe-DevTools.

Reviewer 5: @kemister85 Yes, I have tested and locally with the accessibility tools, it doesn't flag as an error.

> In {productname} 6.4.1, when the sidebar is either visible or hidden it now passes the accessibility test with Axe-DevTools.

Author: This has been updated- closing this thread.

As the paragraph tag is intended to be a block for `text`, it is classified as `semantically incorrect` to contain other block elements.
In this particular instance, child elements in {productname} such as the `<p>` tag would split the block element resulting in a new block (such as a new paragraph) to be created inside the paragraph when a user attempted to edit it inside a `contenteditable="false"` root element. 

Fixes were introduced with the release of {productname} 6.4.1, that:

* Prevents editing host elements like `<p contenteditable="true">paragraph text</p>` in a `contenteditable="false"` root to be split.
* Prevents editing hosts elements inside a `contenteditable="false"` root to be split.

As a result, when a root element is set to `contenteditable="false"`, any `contenteditable="true"` elements inside that root element such as `<p>` tag will no longer try to split the `contenteditable="true"` element. 

=== Creating a list in a table cell when the caret is in front of an anchor element would not properly include the anchor in the list.
//#TINY-6853

> In previous versions of {productname}, a bug was causing issues with `links` while they were nested within table-cells. As a consequence, it was found that when creating a `num-list` or `bullet-list` to the left (node) of a `link` that was nested within a table cell the `list` would *not* include the created `link` within the `num or bullet-list` as expected. 
> 
> To fix this bug, the table cell now checks *If* the caret position is in front of an *anchor element*, `if` (true), then the table cell will include the anchor in the `list`.
> 
> In {productname} 6.4.1, anchor elements are included inside `lists` while nested within table cells.

Reviewer 2: ```suggestion
In previous versions of {productname}, when the caret is to the left of an anchor element at the start of a line in a table cell, creating a `<ul>` or `<ol>` list would not include the anchor element. Instead, the line with the anchor element would be unexpectedly pushed below and an empty list would be created above.

To resolve this bug, in {productname} 6.4.1, when toggling a list in a table cell, inline nodes such as anchors to the left and right of the caret are traversed to ensure that they are included in the list as expected.
```

In {productname} 6.4.1, pressing backspace after adding a Text Pattern, now formats the content as expected.

=== Incorrect event target for dragend and drop events when dragging and dropping noneditable elements within editor
//#TINY-9599

A bug was discovered in previous version of {productname}, that would arise when dragging and dropping `contenteditable="false"` elements within the editor was that the target of the dragstart, dragend or drop events would be the editor instance itself, rather than the expected DOM target.

In {productname} 6.4.1, the events are now simulated based on the mouse event that produced them, meaning that the target and other properties, such as mouse coordinates, are now properly produced for the simulated drag and drop events.

> === Dialog getting cut offscreen when opened from toolbar button in Inline Mode
> //#TINY-9588
> 
> In previous versions of {productname}, inline dialogs would always open in the South West layout from the toolbar, or South East with RTL (Right-to-Left) directionality. However, this sometimes caused problems when the toolbar was shorter than the inline dialog, causing the dialog to open partially off screen.
> 
> To address this issue, improvements made allow inline dialogs the flexability to be opened in other South layouts when they would otherwise be outside of the viewable area. Moreover, inline dialogs now anchor to the toolbar in an appropriate direction with the entire dialog on screen when opened, ensuring a better user experience.

Reviewer 4: ```suggestion
=== Dialog getting cut offscreen when opened from toolbar button in Inline Mode
//#TINY-9588

In previous versions of {productname}, inline dialogs would always open to the South West of the button used to open it from the toolbar, or South East with RTL (Right-to-Left) directionality. However, an issue was found where, when the toolbar was shorter than the inline dialog, inline dialogs would appear partially off screen when opened.

To address this issue, inline dialogs now determine a more appropriate direction to open, that is, any direction which will not cause the dialog to appear offscreen.
```

> === Checkmark icons are added to every color in color swatch
> //#TINY-9567
> 
> In previous versions of {productname}, the color swatch feature added a checkmark icon to every color, but would only display the checkmark when one is enabled.
> 
> In {productname} 6.4.1, the selector for applying the checkmark is updated so the checkmark will only be added and displayed for enabled colors.

Reviewer 4: ```suggestion
=== Checkmark icons are added to every color in color swatch
//#TINY-9567

In previous versions of {productname}, when using certain configurations, the color swatch feature could add a checkmark icon to every color, when the checkmark should only be displayed on a color when one is enabled.

In {productname} 6.4.1, the selector for applying the checkmark is updated so the checkmark will only be added and displayed for enabled colors.
```

=== Plugin - Search and Replace - Throws a JS console error if you search for a string that doesn't exist and try to click on the preferences button.
//#TINY-9686

In previous versions of {productname}, the 'Search and Replace' plugin had a known issue that caused a console error when searching for a string that wasn't in the editor content, and then attempting to click on the `preferences` button.

As a result, an "incorrect type" error would prevent users from opening the preferences section.

In {productname} 6.4.1, this issue has been resolved.

=== Toolbar doesn't hide when editor is out of view in scrollable container
//#TINY-9652

In previous versions of {productname}, when the editor is within a scrollable container, the toolbar would remain visible when the editor is scrolled out of view.

With the release of {productname} 6.4.1, the new `ui_mode` option provides improvements to editors within scrollable contexts, including hiding the toolbar when the editor is out of view when the `ui_mode` option is set to 'split'.

> === `color_cols`: 0 Memory Leak

Reviewer 4: ```suggestion
=== `color_cols`: 0 caused the page to crash
```
Just used memory leak in the name of the ticket, this is probably more suitable

  advcode_inline: true,
  plugins: [
    "code", "fullscreen", 'advcode',
  ],
  toolbar: "code"
});
----

As a result, the toggle button can be found in the top left corner of the Advanced Code Editor Premium toolbar. Once clicked, it will expand the dialog toolbar to fill the entire window, providing a more comprehensive view of the available options. A second click of the toggle button will return the dialog toolbar to its original size.

This option is now available in {productname} 6.4.1, as part of the core/ui API.

=== New text-size-increase and text-size-decrease icons
//#TINY-9530

The {productname} package files have been updated with two new icons that enable users to increase or decrease the size of text in the Advanced Code Editor. The new icons are the `text-size-increase` and `text-size-decrease` icons. These new icons have been added to make it easier for users to adjust the size of the text within the code editor while making it easier to read and understand the code they are writing. 

.Example of `textSizeIncrease` and `textSizeDecrease` integration within {productname}.
[source, js]
----
tinymce.init({
  selector: "textarea",
  advcode_inline: true,

>   plugins: [
>     "code", 'advcode', 
>   ],

Reviewer 6: ```suggestion
  plugins: [
    "code", "advcode", 
  ],
```

//#TINY-9499

In previous versions of {productname}, a formatting issue was found while working with checklists on formatted lists and unordered lists. This issue resulted in incorrect formatting of certain elements when a selection contained a mix of list elements.

In {productname} 6.4.1, this issue has been addressed.

Applying checklists on a selection that includes lists now converts the selected items into checklists, as expected.

For information on the **Checklist** plugin, see: xref:checklist.adoc[Checklist].

=== Merge Tags 1.2.1

The {productname} 6.4.1 release includes an accompanying release of the **Merge Tags** premium plugin.

**Merge Tags** 1.2.1 includes the following change.

==== The mceMergeTag command now preserves the selected content as a plain text and wraps it in a merge tag prefix and suffix if the content length exceeds the limit.
//#TINY-9444

In previous versions of {productname}, the prefix and suffix were removed when merge tag content exceeded the character limit for the merge tag.

In {productname} 6.4.1 this content is retained, as well as the prefix and suffix, though it is not converted into a Merge Tag.

[NOTE]

> Any content exceeding the character limit for Merge Tags will be converted into a Merge Tag.

Reviewer 7: ```suggestion
Any content exceeding the character limit for Merge Tags will not be converted into a Merge Tag.
```

* xref:new-premium-plugin[New Premium Plugin]
* xref:accompanying-premium-plugin-changes[Accompanying Premium Plugin changes]
* xref:accompanying-premium-plugin-end-of-life-announcement[Accompanying Premium Plugin end-of-life announcement]
* xref:accompanying-premium-skins-and-icon-packs-changes[Accompanying Premium Skins and Icon Packs changes]
* xref:improvement[Improvement]
* xref:additions[Additions]
* xref:changes[Changes]
* xref:bug-fixes[Bug fixes]
* xref:known-issues[Known issues]
// NB: edit the equivalent section of nav.adoc when the remove or make visible decision is made.

NOTE: Although there was a Community release of {productname} 6.4, it was not released as a {enterpriseversion} or {cloudname} release. The first commercially supported release of {productname} post version 6.3.2 is {productname} 6.4.1.

[[new-premium-plugin]]
== New Premium Plugin

=== Advanced Template 1.1.0

The {productname} 6.4.1 release includes the release of a new premium plugin: **Advanced Template**.

The **Advanced Template** Premium plugin enables the creation of complex, interactive templates using {productname}.

> Templates are declared as `const` variables that can be setup as part of a {productname} instance.

Reviewer 7: Remove this line.
```suggestion
```

They can also be added to a {productname} instance by end-users selecting content in a {productname} document and saving the selection as a template via the {productname} user-interface.

For information on the **Advanced Template** plugin, see: xref:advanced-templates.adoc[Advanced Template].

[[accompanying-premium-plugin-changes]]
== Accompanying Premium Plugin changes

The following premium plugin updates were released alongside {productname} 6.4.1.

> === Advanced Code Editor 3.2.1
> 
> The {productname} 6.4.1 release includes an accompanying release of the **Advanced Code Editor** premium plugin.
> 
> **Advanced Code Editor** 3.2.1 includes the following improvements and additions.
> 
> ==== New button to copy the current dialog content.
> 
> This update adds a new **Copy code** button to the Advanced Code Editor dialog, enabling the copying of the entire contents of the dialog without having to first select.
> 
> ==== New buttons to increase and decrease the font size in the dialog.
> 
> This update adds two new buttons for increasing and decreasing the font size of code displayed in the Advanced Code Editor dialogs.
> 
> ==== New button to toggle dark mode.
> 
> This update adds a new **Dark/light mode** button for toggling the Advanced Code Editor dialog between dark and light display modes.
> 
> ==== New button to toggle fullscreen.
> 
> This update adds a new **Fullscreen** button that for toggling the Advanced Code Editor dialog between fullscreen and dialog modes.
> 
> NOTE: For the Advanced Code Editor to offer a full-screen mode requires the xref:fullscreen.adoc[Full screen] plugin and requires {pluginname} to be running in xref:advcode.adoc#advcode_inline[inline mode].

Reviewer 8: ```suggestion
==== New button to copy the current Code Editor content.

This update adds a new **Copy code** button to the Advanced Code Editor Code Editor, enabling the copying of the entire contents of the Code Editor without having to first select.

==== New buttons to increase and decrease the font size in the Code Editor.

This update adds two new buttons for increasing and decreasing the font size of code displayed in the Advanced Code Editor Code Editors.

==== New button to toggle dark mode.

This update adds a new **Dark/light mode** button for toggling the Advanced Code Editor Code Editor between dark and light display modes.

==== New button to toggle fullscreen.

This update adds a new **Fullscreen** button that for toggling the Advanced Code Editor Code Editor between fullscreen and Code Editor initial size.

NOTE: For the Advanced Code Editor to offer a full-screen mode requires the xref:fullscreen.adoc[Full screen] plugin and requires {pluginname} to be running in xref:advcode.adoc#advcode_inline[inline mode].

NOTE: For the Advanced Code Editor to offer a Copy button requires a https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts[*secure contexts*]
```

If something sounds bad in English feel free to change it, I just removed the term `dialog` and used `Code Editor` to refer to both dialog and inline mode

This is particularly important for the fullscreen mode which can be used only in inline mode.

I also add a second `NOTE` if it is not correct to have 2 notes in row, feel free to merge them

Author: ```suggestion
=== Advanced Code Editor 3.2.1

The {productname} 6.4.1 release includes an accompanying release of the **Advanced Code Editor** premium plugin.

**Advanced Code Editor** 3.2.1 includes the following improvements and additions.

==== New button to copy the current dialog content.

This update adds a new **Copy code** button to the **Advanced Code Editor** for editors using either `advcode_inline`, `inline` or using the default dialog within {productname}'s `Classic mode`. Its worth noting that both options allow the user to copy the entire contents of from the `source code` without having to first select the content.

==== New buttons to increase and decrease the font size in the dialog.

This update adds two new buttons for increasing and decreasing the font size of code displayed in the Advanced Code Editor.

==== New button to toggle dark mode.

This update adds a new **Dark/light mode** button that can be used within the Advanced Code Editor to toggle between either dark or light display modes.

==== New button to toggle fullscreen.

This update adds a new **Fullscreen** button that can be used within the Advanced Code Editor to toggle between fullscreen and dialog modes.

NOTE: For the Advanced Code Editor to offer a full-screen mode, it requires the xref:fullscreen.adoc[Full screen] plugin and requires {pluginname} to be running in xref:advcode.adoc#advcode_inline[inline mode].

NOTE: For the Advanced Code Editor to offer a **Copy** button, it requires a https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts[*secure contexts*]
```

Author: Discussed with @tiny-ben-tran for approval

=== Checklist 2.0.4

The {productname} 6.4.1 release includes an accompanying release of the **Checklist** premium plugin.

**Checklist** 2.0.4 includes the following fix.

==== Formatting Issue with Checklist on Formatted Lists and Non-Lists
//#TINY-9499

In previous versions of {productname}, a formatting issue was found while working with checklists on formatted lists and unordered lists. This issue resulted in incorrect formatting of certain elements when a selection contained a mix of list elements.

In {productname} 6.4.1, this issue has been addressed.

Applying checklists on a selection that includes lists now converts the selected items into checklists, as expected.

For information on the **Checklist** plugin, see: xref:checklist.adoc[Checklist].

=== Merge Tags 1.2.1

The {productname} 6.4.1 release includes an accompanying release of the **Merge Tags** premium plugin.

**Merge Tags** 1.2.1 includes the following change.

==== The mceMergeTag command now preserves the selected content as a plain text and wraps it in a merge tag prefix and suffix if the content length exceeds the limit.
//#TINY-9444

> In previous versions of {productname}, content which exceeded the character limit for Merge Tags was removed when surrounded by the configured Merge Tag prefix and suffix.

Reviewer 7: ```suggestion
In previous versions of {productname}, the prefix and suffix were removed when merge tag content exceeded the character limit for the merge tag.
```

This new tree component is an important addition to {productname} which will be used in several features with up and coming releases in the future.

=== renderUI property in the Theme type can now return a Promise <RenderResult> instead of RenderResult
//#TINY-9556

In earlier versions of {productname} the `renderUI` function which initiates theme rendering did not wait for the CSS to load. This could cause race conditions with other parts of the editor or customer integrations. As a result {productname} 6.4.1 now waits for a CSS load event before returning from the `renderUI` function. This process necessitates changing the return type of the `renderUI` function; it now returns a `Promise<RenderResult>`. The type is `Promise<RenderResult> | RenderResult` for compatibility, but the `silver` theme always returns a promise.

This change should have no impact on customer integrations, even those implementing custom themes.

> === New `editor.selection.isEditable` API method that returns whether the current selection is editable
> //#TINY-9462
> 
> In previous versions of {productname}, when the body was set to `contenteditable="false"` content could still be inserted with the `insertContent` API.
> 
> Example content:
> [source, HTML]
> ----
> <body contenteditable="false"><p>Try to insert content here</p></body>
> ----
> 
> After selecting the space between "insert" and "content", executing `tinymce.activeEditor.insertContent('text')` would incorrectly insert `text` at the selection which should not be possible.
> 
> [source, HTML]
> ----
> <body contenteditable="false"><p>Try to inserttextcontent here</p></body>
> ----
> 
> {productname} 6.4 includes a new `isEditable` method on both the `editor.selection` and `editor.dom` APIs which are used to block the `editor.insertContent` API and `mceInsertContent` command when the selection covers a `contenteditable="false"` element.

Reviewer 9: Since the insertContent blocking is described in it's own section later it might be better to just describe the API here what that does. Hard to describe exactly what it's doing but I gave it a shot here.

```suggestion
=== New `editor.selection.isEditable` API method that returns whether the current selection is editable
//#TINY-9462

{productname} 6.4 includes a new `isEditable` method on the `editor.selection` that will check if the selection  is editable by finding the closest ascendant element for each selection endpoint with either a `true` or `false` value for it's `contenteditable` attribute. If both endpoints of the selection has it's closest ancestor `contenteditable` attribute value of `true` then the selection is editable.
```

=== New `editor.dom.isEditable` API method that returns whether the specified node is editable
//#TINY-9462

In previous versions of {productname}, elements within a `contenteditable="false"` root could be unexpectedly modified by the `editor.insertContent` API and the `mceInsertContent` command..

As a consequence all elements within the root `body` set to `contenteditable="false";` value, were affected by the issue.

> The new additions and improvements with the introduction of {productname} 6.4 include new `editor.dom.isEditable` and `editor.selection.isEditable` APIs which now block the `editor.insertContent` API and `mceInsertContent` command if the selection is in a `contenteditable="false"` part of the root contents.

Author: ```suggestion
{productname} 6.4.1 includes a new `isEditable` method on both the `editor.selection` and `editor.dom` APIs which are used to block the `editor.insertContent` API and `mceInsertContent` command when the selection covers a `contenteditable="false"` element.

=== New `editor.dom.isEditable` API method that returns whether the specified node is editable
//#TINY-9462

In previous versions of {productname}, elements within a `contenteditable="false"` root could be unexpectedly modified by the `editor.insertContent` API and the `mceInsertContent` command..

As a consequence all elements within the root `body` set to `contenteditable="false";` value, were affected by the issue.

The new additions and improvements with the introduction of {productname} 6.4.1 include new `editor.dom.isEditable` and `editor.selection.isEditable` APIs which now block the `editor.insertContent` API and `mceInsertContent` command if the selection is in a `contenteditable="false"` part of the root contents.
```

* For Method 3: When one of the keyboard shortcuts from Method 3 is detected on `keyup`, the override creates an unformatted caret at the current selection. Performing the override on `keyup` allows the native behavior of the deletion shortcuts to occur. Then, applying an unformatted caret at the selection after deletion will prevent the browser from recreating any formats.

=== Visual characters were rendered inside noneditable elements
//#TINY-9474

In previous versions of {productname}, a bug was found that affected the `visual characters` plugin. The visual characters would be rendered within an `contenteditable="false"` element.

In order to address this problem, improvements were made to better detect noneditable context, when an element is identified as a `contenteditable="false"` element, either by itself or through its descendants.

As a result, in {productname} 6.4.1, visual characters no longer render inside `contenteditable="false"` elements.

=== Lists in a noneditable root were incorrectly editable using list API commands, toolbar buttons and menu items
//#TINY-9458

> In previous versions of {productname}, a bug was found that affected the *lists*  and *Advanced List* plugins when the {productname} editor body was set to `contenteditable = "false"`. The consequence meant that the lists, that were even set as `contenteditable="false"`,  elements could be modified by using list API commands, toolbar buttons and menu items.

Author: @spocke Its old habit for adding them 😆

//#TINY-9459

In versions of {productname}, prior to 6.4.1, tables could be modified by using `execCommands` when tables are placed within a `contenteditable="false"` root element.

As a consequence, a bug was identified that by using the `execCommands`, which are provided by {productname}, the user could modify the table element regardless of it being nested within a `contenteditable="false"` root element. This was made possible by by overriding the `contenteditable="false"` element settings by using the `execCommand`'s' to trigger the `action` event.

To fix this bug, {productname} 6.4.1, made improvements to the DOM `Commands` API that now blocks all Table UI actions via the `execCommands` when a `contenteditable="false"` element selection is highlighted inside a `contenteditable="false"` root.

As a result, `contenteditable="false"` elements such as tables cannot have their properties overridden by using the `execCommands`.

_(Readmore about editorCommands xref:editor-command-identifiers.adoc[here])_

=== Fake carets were rendered for noneditable elements and tables in a noneditable root element
//#TINY-9459

> In previous versions of {productname}, `fake` carets that are within a `contenteditable="false"` element, would render as a `highlighted` node, when nested within another `contenteditable="false"` element, this would occured when the content was placed within a `table` placed in the root element.
> 
> As a consequence, `fake` caret/s would be visibly highlighted within the editor viewport,  which would allow content to be inserted into the element when the table was selected by using either the keyboard and mouse function commands.
> 
> To fix this {productname} 6.4.1, added a new function to the Utils API that filters any content that is considered as a `fake caret` element, which now disables the users ability to add/insert or modify the content contained within the `contenteditable="false"` element node space.
> 
> As a result, any `contenteditable="false"` elements that are nested within another `contenteditable="false"` element root that are nested within `table` elements, can no longer be edited in {productname} 6.4.1.

Author: ```suggestion
In previous versions of {productname}, `fake` carets would be rendered before or after tables or `contenteditable="false"` elements that was nested within another `contenteditable="false"` element.

To fix this {productname} 6.4.1, added a blocker for rendering fake carets for `contenteditable="false"` and table elements that are contained within other `contenteditable="false"` elements.

As a result, any `contenteditable="false"` or table elements that are nested within another `contenteditable="false"` element root will no longer have fake carets in {productname} 6.4.1.
```

As a result, in {productname} 6.4.1, when using the backspace or delete function keys on `contenteditable="false"` elements within a `contenteditable="false"` element, the contents within the `contenteditable="false"` elements will no longer be deleted. 

=== Inserting newlines inside an editable element inside a noneditable root would sometimes try to split the editable element
//#TINY-9461

In previous versions of {productname}, inserting newlines inside an `contenteditable="true"` element while it was inside a `contenteditable="false"` root would sometimes try to split the `contenteditable="true"` element. 

As a consequence, when child block elements set to `contenteditable="true"` such as a `<p>` tag were nested within the `contenteditable="false"` element root, the editor would try to split the child block element when the user hit the `enter` command. 

> As the paragraph tag is intended to be a block for `text`, it is classified as `semantically incorrect` to split a paragraph block element.
> In this particular instance, child elements in {productname} such as the `<p>` tag would split the block element resulting in a new block (such as a new paragraph) to be created inside the paragraph when a user attempted to edit it inside a `contenteditable="false"` root element. 
> 
> Fixes were introduced with the release of {productname} 6.4.1, that:
> 
> * Prevents editing host elements like `<p contenteditable="true">paragraph text</p>` in a `contenteditable="false"` root to be split.
> * Prevents editing hosts elements inside a `contenteditable="false"` root to be split.
> 
> As a result, when a root element is set to `contenteditable="false"`, any `contenteditable="true"` elements inside that root element such as `<p>` tag will no longer try to split the `contenteditable="true"` element. 

Reviewer 9: ```suggestion
As the paragraph tag is intended to be a block for `text`, it is classified as `semantically incorrect` to contain other block elements.
In this particular instance, child elements in {productname} such as the `<p>` tag would split the block element resulting in a new block (such as a new paragraph) to be created inside the paragraph when a user attempted to edit it inside a `contenteditable="false"` root element. 

Fixes were introduced with the release of {productname} 6.4.1, that:

* Prevents editing host elements like `<p contenteditable="true">paragraph text</p>` in a `contenteditable="false"` root to be split.
* Prevents editing hosts elements inside a `contenteditable="false"` root to be split.

As a result, when a root element is set to `contenteditable="false"`, any `contenteditable="true"` elements inside that root element such as `<p>` tag will no longer try to split the `contenteditable="true"` element. 
```

//#TINY-6853

In previous versions of {productname}, a bug was causing issues with `links` while they were nested within table-cells. As a consequence, it was found that when creating a `num-list` or `bullet-list` to the left (node) of a `link` that was nested within a table cell the `list` would *not* include the created `link` within the `num or bullet-list` as expected. 

To fix this bug, the table cell now checks *If* the caret position is in front of an *anchor element*, `if` (true), then the table cell will include the anchor in the `list`.

In {productname} 6.4.1, anchor elements are included inside `lists` while nested within table cells.

=== Dragging and dropping noneditable elements on table borders would remove the element on drop
//#TINY-9021

In previous versions of {productname}, dragging a `contenteditable="false"` element and dropping it on a table border would cause the `contenteditable="false"` element to be deleted. The user often intends to drop the element into a table cell, but may accidentally drop it on a table border, undesirably causing the element's deletion.

> {productname} 6.4.1 introduced a new `isActive` api for Dragster that checks whether an element is being dragged or not. If an element is being dragged, the resize bars on table borders are hidden, which prevents the deletion side-effect of dropping elements onto the table border resize bars. If an element is dropped onto a table border, it will now be inserted into the left or right adjacent cell instead of being deleted.

Reviewer 9: Removed references to exact implementation and internal libraries.

```suggestion
{productname} 6.4.1 introduced a new behavior if an element is dropped onto a table border, it will now be inserted into the left or right adjacent cell instead of being deleted.
```

In {productname} 6.4.1, pasting blocks that are invalid children of a `contenteditable="true"` blocks will no longer split that block, but instead unwrap the blocks to fit within the specific `contenteditable="true"` block.

=== color_cols option was not respected in the forecolor or backcolor color swatches
//#TINY-9560

In previous versions of {productname}, the logic for specifying the number of columns in the `forecolor` and `backcolor` swatches was updated with the release of {productname} 6.3.

As a consequence, this logic didn't use the color_cols option in {productname} v6.3

[source, js]
----
tinymce.init({
    selector: "textarea",
    plugins: "table code",
    toolbar: "forecolor backcolor tableprops",
    color_cols: 7, // <-- option was not respected
});
----

In {productname} 6.4.1, the color swatches opened by the `forecolor` and `backcolor` toolbar buttons have the amount of columns set by the `color_cols` option, if it is set.

=== Drag and dropping the last noneditable element out of its parent block would not properly pad the parent block element
//#TINY-9606

> In previous versions of {productname}, dragging and dropping `contenteditable="false"` elements into another block elements would cause issues with the original parent block. As a consequence, the parent block element where the `contenteditable="false"` element was dragged from would leave the parent block as a `contenteditable="false"` element, as it was not discarding the contents when the element was transferred.
> 
> As a fix, {productname} 6.4.1, added a new condition which validates a set of conditions the check if the parent block element is within a `parentNode` and is a block element, in which it will trigger a action to remove elements with padding that meet this condition. 
> 
> In {productname} 6.4.1, dragging and dropping elements that are `contenteditable="false"` out of they're parent blocks now correctly pad the parent block element.

Reviewer 9: Maybe something like:

```suggestion
In previous versions of {productname}, dragging the last child element with a `contentedtiable="false"` attribute out of a block element would prevent the user from placing the caret back into that source block.

As a fix, {productname} 6.4.1, added a new condition which validates if the source element is empty when the drop operation occurs. If it is empty then it will be padded with a `BR` element so that it's possible to place the caret inside the block. 

In {productname} 6.4.1, dragging and dropping elements that are `contenteditable="false"` out of they're parent blocks now correctly pad the parent block element.
```

To fix this issue, a workaround was implemented in {productname} 6.4.1, that now bypasses the default paste functionality for this undo function.

In {productname} 6.4.1, pasting links to text selections now generates the correct undo stack on Safari.

=== Toolbar split buttons in *Advanced List* plugin to show the correct state when the cursor is in a checklist
//#TINY-5167

In previous versions of {productname}, when a cursor was placed within a checklist item, the icon for both the checklist and the bullet list would show as `active`. This created UX issues, as having both buttons showing an `active` state was confusing for the user, specifically for navigation purposes when using *Advanced List* plugin and checklists from the *Checklist* plugin.

To fix this issue, {productname} updates the logic to prevent setting the bullet list to `active` when the cursor is in a checklist.

As a result, in {productname} 6.4.1, only the checklist `icon` will appear as `active` when the cursor is in a checklist.

>  (edited) 

Reviewer 6: ```suggestion
```

//#TINY-9600

Currently, we use the DOMPurify library to sanitize HTML content when it is inserted into the editor in order to remove unsafe tags & attributes in order to prevent XSS (cross-site scripting). However, customers in some situations have found that the HTML sanitization from the DOMPurify library is changing their content in undesired ways. 

To address this issue, we have added a new `xss_sanitization: boolean` editor option. This option is enabled by default which enables sanitization and can be set to `false` to disable sanitization. If a customer wishes to prevent their content from being modified by the sanitization, they can now set this option to `false`.

.Example: using `+xss_sanitization+` to turn DOMPurify off and no longer sanitize HTML against XSS attacks
[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your HTML
  xss_sanitization: false
});
----

[IMPORTANT]
It is important to note that disabling sanitization introduces potential security issues, especially if the server-side XSS sanitization is weak or non-existent. Using this option therefore not recommended.

> // Read more about xref:sanitizing-html-input-to-protect-against-xss-attacks[Sanitizing HTML input to protect against XSS attacks]

Author: ```suggestion
Read more about xref:sanitizing-html-input-to-protect-against-xss-attacks[Sanitizing HTML input to protect against XSS attacks]
```

=== Toolbar split buttons in *Advanced List* plugin to show the correct state when the cursor is in a checklist
//#TINY-5167

In previous versions of {productname}, when a cursor was placed within a checklist item, the icon for both the checklist and the bullet list would show as `active`. This created UX issues, as having both buttons showing an `active` state was confusing for the user, specifically for navigation purposes when using checklists from the *Advanced List* plugin.

To fix this issue, {productname} added improvements to the list configuration to checks a new set of conditionals which would determine if the `node` selected matched the parent `node` within the checklist. If the match returns as `true`, the (api) will set only the checklist icon to a `active` state.

As a result, the checklist `icon` in {productname} 6.4.1, related to the `advlist` plugin will only appear in a `active` state, when the `node` checklist meets the new conditions implemented.

=== Dragging transparent elements into transparent blocks elements could produce invalid nesting of transparents
//#TINY-9231

> Dragging or dropping block `links` into other `links` was producing invalid nesting of other transparents such as `anchors`. As a result, these transparent elements would produce `anchor` tags within other `anchor` tags, which is not a general practice. As an example `links` and `anchors` that are defined to an element, must not be nested within or contain any other element and vice versa.
> 
> Changes were implemented in {productname} 6.4.1 to fix this issue to ensure that any invalid child anchor elements that try to reproduce this action are unwrapped when the editor verifies another transparent has been dropped into another transparent (anchor) element.
> 
> In {productname} 6.4.1, dragging content such as `links` into other `links` will now produce *valid* HTML `output` after a drag operation occurs which will be consistent across all browsers.

Reviewer 2: ```suggestion
In previous versions of {productname}, dragging and dropping inline anchor links into block anchor links was producing invalid nesting of anchors in Chrome. Anchors, like all other transparent elements, must not be nested within each other.

Changes were implemented in {productname} 6.4.1 such that when an inline transparent element such as an anchor is dragged and dropped into a transparent block element of the same type (e.g. also an anchor), the dropped inline transparent element is unwrapped to prevent the nesting behavior. This ensures that the HTML output after the drag operation is valid and consistent across browsers.
```

//#TINY-9459

In previous versions of {productname}, `fake` carets that are within a `contenteditable="false"` element, would render as a `highlighted` node, when nested within another `contenteditable="false"` element, this would occured when the content was placed within a `table` placed in the root element.

As a consequence, `fake` caret/s would be visibly highlighted within the editor viewport,  which would allow content to be inserted into the element when the table was selected by using either the keyboard and mouse function commands.

To fix this {productname} 6.4.1, added a new function to the Utils API that filters any content that is considered as a `fake caret` element, which now disables the users ability to add/insert or modify the content contained within the `contenteditable="false"` element node space.

As a result, any `contenteditable="false"` elements that are nested within another `contenteditable="false"` element root that are nested within `table` elements, can no longer be edited in {productname} 6.4.1.

=== Textareas with scrollbars in dialogs would not render rounded corners correctly on some browsers
//#TINY-9331

> In previous versions of {productname}, the source code visible UI border would not render correctly in some browsers.
> 
> As a consequence, when viewing text-areas that contain `scrollbars` within dialogs, the `focus` visible border was not rendering the corners inside the dialog.
> 
> To correct this issue, {productname} 6.4.1, introduced a new text-area wrapper that fixes the issues with the `border-color, radius, style and width` that directly affects the `focus` render element for the dialog.  In addition `overflow: hidden`, has been added to clip the scrollbars to the border-radius.

Reviewer 2: ```suggestion
In previous versions of {productname}, textareas with scrollbars in dialogs would not render rounded corners correctly on some browsers, as the rounded corner would be overlaid by the right-angled corner of the scrollbar track. An example of a dialog affected by this issue was the source code dialog.

{productname} 6.4.1 added to the form textarea element a wrapper class containing border properties and an `overflow: hidden` property. This correctly clips the scrollbars to the border-radius on all browsers, resolving the bug.
```

As a result, the `focus` dialog border now renders correctly in {productname} 6.4.1.

=== It was possible to open links inside the editor if the editor root was noneditable
//#TINY-9470
While setting the editor root to `contenteditable="false"`, it was discovered that links could still be opened, even in instances where the editor was setup as the below:
.Example setup
[source, js]
----
tinymce.init({
    selector: "textarea",

>     plugins: [
>     ],

Reviewer 6: ```suggestion
```

=== Dialog getting cut offscreen when opened from toolbar button in Inline Mode
//#TINY-9588

In previous versions of {productname}, inline dialogs would always open in the South West layout from the toolbar, or South East with RTL (Right-to-Left) directionality. However, this sometimes caused problems when the toolbar was shorter than the inline dialog, causing the dialog to open partially off screen.

To address this issue, improvements made allow inline dialogs the flexability to be opened in other South layouts when they would otherwise be outside of the viewable area. Moreover, inline dialogs now anchor to the toolbar in an appropriate direction with the entire dialog on screen when opened, ensuring a better user experience.

=== Checkmark icons are added to every color in color swatch
//#TINY-9567

In previous versions of {productname}, the color swatch feature would add a checkmark icon to every color, but would only display the checkmark on the "aria-checked" attribute.

> In {productname} 6.4.1, a fix was applied that replaced the previous selector, `aria-checked='true'``, with `tox-collection__item--enabled`. This fix, ensures that the checkmark is only displayed on the enabled colors.

Reviewer 6: ```suggestion
In previous versions of {productname}, the color swatch feature added a checkmark icon to every color, but would only display the checkmark when one is enabled.

In {productname} 6.4.1, the selector for applying the checkmark is updated so the checkmark will only be added and displayed for enabled colors.

This new tree component is an important addition to {productname} which will be used in several features with up and coming releases in the future.

=== renderUI property in the Theme type can now return a Promise <RenderResult> instead of RenderResult
//#TINY-9556

In earlier versions of {productname} the `renderUI` function which initiates theme rendering did not wait for the CSS to load. This could cause race conditions with other parts of the editor or customer integrations. As a result {productname} 6.4.1 now waits for a CSS load event before returning from the `renderUI` function. This process necessitates changing the return type of the `renderUI` function; it now returns a `Promise<RenderResult>`. The type is `Promise<RenderResult> | RenderResult` for compatibility, but the `silver` theme always returns a promise.

This change should have no impact on customer integrations, even those implementing custom themes.

> === New isEditable API to editor.selection that returns true or false if the current selection is editable

Reviewer 2: ```suggestion
=== New `editor.selection.isEditable` API that returns whether the current selection is editable
```

//#TINY-9462

In previous versions of {productname}, when the body was set to `contenteditable="false"` content could still be inserted with the `insertContent` API.

Example content:
[source, HTML]
----
<body contenteditable="false"><p>Try to insert content here</p></body>
----

> After selecting the space between "insert" and "content", executing `tinymce.activeEditor.insertContent('text')` would incorrectly insert the text which should not be possible.

Reviewer 2: ```suggestion
After selecting the space between "insert" and "content", executing `tinymce.activeEditor.insertContent('text')` would incorrectly insert `text` at the selection which should not be possible.
```

[source, HTML]
----
<body contenteditable="false"><p>Try to inserttextcontent here</p></body>
----

{productname} 6.4 includes a new `isEditable` method on both the `editor.selection` and `editor.dom` APIs which are used to block the `editor.insertContent` API and `mceInsertContent` command when the selection covers a `contenteditable="false"` element.

> === New `editor.dom.isEditable` and `editor.selection.isEditable` APIs that returns whether the specified node is editable

Reviewer 2: ```suggestion
=== New `editor.dom.isEditable` API that returns whether the specified node is editable
```

==== Lists separated by empty paragraphs resulted in incorrect numbering when pasted from Microsoft Word
//#TINY-6446

Previously, when an ordered list, with nested elements, that also included blank lines between list items was copied from Microsoft Word and pasted into a {productname} editor instance, the numbering of the pasted-in list was incorrect.

In pasting, the PowerPaste plugin made two errors:

. it incorrectly incremented outer list items where these increments should have only applied to nested items; and
. it incorrectly joined the nested list item numbering.

This update corrects both these errors. Pasting lists from Microsoft Word into a {productname} editor now shows the correct numbering in the pasted list, even when the list includes non-list paragraphs between list items.

For information on the **PowerPaste** plugin, see: xref:introduction-to-powerpaste.adoc[PowerPaste].

=== Comments 3.3.1

The {productname} 6.4.1 release includes an accompanying release of the **Comments** premium plugin.

**Comments** 3.3.1 includes the following fix.

==== The comment input box now has a max height to prevent long comments from expanding the box indefinitely
//#TINY-8894

> In previous versions of {productname}, the comment input box for the popup dialog would exceed the max height for the {productname} viewport.
> 
> As a consequence, in situations when the user would insert content into the input field, the height of the comment box would expand beyond the max viewport height (vh), for {productname}.  This caused UX issues which restricted the users ability to `submit` the add **Comment** button as it was forced outside of the border `bounds` of the {productname} editor. 
> 
> To fix this issue, {productname} introduced two changes which improved the way the **Comment** input box behaves.
> 
> . Updates to the max height for the **Comment** input box, which now prevents long comments from expanding the box indefinitely beyond the editor set `bounds`. 
> . Set new heights for the **Comment** containers that is focused on improving the visibility of long comments by containing them within a smaller scrolling space with the added benefit of `SHOW MORE` or `SHOW LESS`.
> 
> As a result, the input box for the **Comments** premium plugin now offers these improvements with the most recent release of {productname} 6.4.1.

Reviewer 2: ```suggestion
In previous versions of {productname}, if the comment string is larger than the initial height of the comment box, the comment input box would automatically expand its height to fit the comment string. However, when given long enough comment strings, the comment input box would grow beyond the viewable bounds of {productname}. This caused UX issues as:

* the out-of-bounds parts of the comment was not viewable or editable; and 
* the user was unable to submit the comment as the Add Comment button was also forced out of bounds.

To address this issue, the **Comments** 3.3.1 in {productname} 6.4.1 set a maximum height for the comment input box, defined as half the height of the conversations area in the sidebar. At this maximum height, the height of the comment input box will be the same height as the viewing area for existing comments above, providing visual symmetry. The comment input box will not grow beyond the maximum height, and if the comment string is larger than this maximum height, the comment input box can be scrolled for the user to view and edit the entire comment string.
```

  plugins: [
    "code", 'advcode', 
  ],
  toolbar: "textSizeIncrease textSizeDecrease"
});
----
[NOTE]
This plugin feature is only available for paid {productname} subscriptions.

=== New xss_sanitization option to allow for XSS sanitization to be disabled
//#TINY-9600

Currently, we use the DOMPurify library to sanitize HTML content when it is inserted into the editor in order to remove unsafe tags & attributes in order to prevent XSS (cross-site scripting). However, customers in some situations have found that the HTML sanitization from the DOMPurify library is changing their content in undesired ways. 

To address this issue, we have added a new `xss_sanitization: boolean` editor option. This option is enabled by default which enables sanitization and can be set to `false` to disable sanitization. If a customer wishes to prevent their content from being modified by the sanitization, they can now set this option to `false`.

.Example: using `+xss_sanitization+` to turn DOMPurify off and no longer sanitize HTML against XSS attacks
[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your HTML
  xss_sanitization: false
});
----

[IMPORTANT]

> It is important to note that disabling sanitization introduces potential security issues, especially if the server-side XSS sanitization is weak or non-existent and is therefore not recommended.

Reviewer 2: ```suggestion
It is important to note that disabling sanitization introduces potential security issues, especially if the server-side XSS sanitization is weak or non-existent. Using this option therefore not recommended.
```

> // Read more about xref:sanitizing-html-input-to-protect-against-xss-attacks[Sanitizing HTML input to protect against XSS attacks]

Reviewer 2: I think this will be a useful link, so perhaps we should uncomment it.

=== The editor.insertContent API would insert contents inside noneditable elements if the selection was inside the element
//#TINY-9462

In previous versions of {productname}, the settings for the `contentEditable` property in the editor root was set to `false`. As a consequence, the user could insert content using `api` commands to edit a `contenteditable="false"` root element.

To fix this bug, in {productname} 6.4.1, a blocker was added to the `editor.insertContent` API and `mceInsertContent` command, which stops `contenteditable="false"` root elements from being edited.

In {productname} 6.4.1, inserting contents inside a `contenteditable="false"` element while the selection is inside another element, is no longer possible.

=== Closing a dialog would scroll down the document in Safari
//#TINY-9148

While using previous versions of {productname}, `clicking` away to dismiss the *Search* and *Replace* dialog within Safari would scroll the page to the bottom of the editor. As a result the user would have to manually scroll back to the original `focus` area each time the steps above were executed.

> To correct this issue, the active element will be blurred before closing the dialog.

Reviewer 5: ```suggestion
To correct this issue, the dialog will be blurred upon closing.
```

Sorry, should've made a more explicit suggestion.

A key override on the Backspace and Delete keys already exists to address the first method. However, the bug persists for the second and third methods.

In {productname} 6.4.1, two new key overrides have been added to address the second and third methods:

* For Method 2: When Backspace or Delete is detected on `keydown` on a ranged selection, the override checks if the conditions for Method 2 is satisfied. If so, the override will read the current formats, delete the selection, and then create a caret containing the original formats.
* For Method 3: When one of the keyboard shortcuts from Method 3 is detected on `keyup`, the override creates an unformatted caret at the current selection. Performing the override on `keyup` allows the native behavior of the deletion shortcuts to occur. Then, applying an unformatted caret at the selection after deletion will prevent the browser from recreating any formats.

=== Visual characters were rendered inside noneditable elements
//#TINY-9474

> In previous versions of {productname}, a bug was found that affected the `visual characters` plugin. As a consequence, the visual characters selector would render inside `contenteditable="false"` elements after the plugin was applied to the editor.

Reviewer 6: ```suggestion
In previous versions of {productname}, a bug was found that affected the `visual characters` plugin. The visual characters would be rendered within an `contenteditable="false"` element.
```

As a result, when a root element is set to `contenteditable="false"`, any `contenteditable="true"` elements inside that root element such as `<p>` tag will no longer try to split the `contenteditable="true"` element. 

=== Creating a list in a table cell when the caret is in front of an anchor element would not properly include the anchor in the list.
//#TINY-6853

In previous versions of {productname}, a bug was causing issues with `links` while they were nested within table-cells. As a consequence, it was found that when creating a `num-list` or `bullet-list` to the left (node) of a `link` that was nested within a table cell the `list` would *not* include the created `link` within the `num or bullet-list` as expected. 

To fix this bug, the table cell now checks *If* the caret position is in front of an *anchor element*, `if` (true), then the table cell will include the anchor in the `list`.

In {productname} 6.4.1, anchor elements are included inside `lists` while nested within table cells.

=== Dragging and dropping noneditable elements on table borders would remove the element on drop
//#TINY-9021

> In previous versions of {productname}, dragging and dropping `contenteditable="false"` elements on table borders would create issues during the drop action.
> 
> As a consequence, this would result in the `contenteditable="false"` element being deleted after the `action` when dropped onto the borderline within the table cell.
> To fix this issue, {productname} 6.4.1, introduced a new `isActive` api to the handler for dragger, that checks if the event `isActive` or not, which then hides the resize bar during the dragging event, which now prevents content from being dropped onto table bars.
> 
> In {productname} 6.4.1, dragging an inline `contenteditable="false"` element onto the border inside a table, now moves it into a neighbouring cell instead of deleting it.

Reviewer 2: ```suggestion
In previous versions of {productname}, dragging a `contenteditable="false"` element and dropping it on a table border would cause the `contenteditable="false"` element to be deleted. The user often intends to drop the element into a table cell, but may accidentally drop it on a table border, undesirably causing the element's deletion.

{productname} 6.4.1 introduced a new `isActive` api for Dragster that checks whether an element is being dragged or not. If an element is being dragged, the resize bars on table borders are hidden, which prevents the deletion side-effect of dropping elements onto the table border resize bars. If an element is dropped onto a table border, it will now be inserted into the left or right adjacent cell instead of being deleted.
```

// NB: edit the equivalent section of nav.adoc when the remove or make visible decision is made.

NOTE: Although there was a Community release of {productname} 6.4, it was not released as a {enterpriseversion} or {cloudname} release. The first commercially supported release of {productname} post version 6.3.2 is {productname} 6.4.1.

[[new-premium-plugin]]
== New Premium Plugin

=== Advanced Template 1.1.0

The {productname} 6.4.1 release includes the release of a new premium plugin: **Advanced Template**.

The **Advanced Template** Premium plugin enables the creation of complex, interactive templates using {productname}.

Templates are declared as `const` variables that can be setup as part of a {productname} instance.

They can also be added to a {productname} instance by end-users selecting content in a {productname} document and saving the selection as a template via the {productname} user-interface.

For information on the **Advanced Template** plugin, see: xref:advanced-templates.adoc[Advanced Template].

[[accompanying-premium-plugin-changes]]
== Accompanying Premium Plugin changes

The following premium plugin updates were released alongside {productname} 6.4.1.

> === Accessibility Checker 3.1.1
> 
> The {productname} 6.4.1 release includes an accompanying release of the **Accessibility Checker** premium plugin.
> 
> **Accessibility Checker** 3.1.1 includes the following addition:
> 
> ==== Updated the **Accessibility Checker** dropdown menu styles to match the default {productname} styles
> //#TINY-9630
> 
> Previously, the **Accessibility Checker** premium plugin used browser-native elements for some dropdown menus.
> 
> With this update, these menus are now created using the equivalent {productname} custom element, making these dropdown menus present with the same styling as other {productname} UI elements.
> 
> For information on the **Accessibility Checker** plugin, see xref:a11ychecker.adoc[Accessibility Checker].
> 
> 

Reviewer 3: This update is intended for 6.5
```suggestion
```

The {productname} 6.4.1 release includes an accompanying release of the **Advanced Code Editor** premium plugin.

**Advanced Code Editor** 3.2.1 includes the following improvements and additions.

==== New button to copy the current dialog content.

This update adds a new **Copy code** button to the Advanced Code Editor dialog, enabling the copying of the entire contents of the dialog without having to first select.

==== New buttons to increase and decrease the font size in the dialog.

This update adds two new buttons for increasing and decreasing the font size of code displayed in the Advanced Code Editor dialogs.

==== New button to toggle dark mode.

This update adds a new **Dark/light mode** button for toggling the Advanced Code Editor dialog between dark and light display modes.

==== New button to toggle fullscreen.

This update adds a new **Fullscreen** button that for toggling the Advanced Code Editor dialog between fullscreen and dialog modes.

NOTE: For the Advanced Code Editor to offer a full-screen mode requires the xref:fullscreen.adoc[Full screen] plugin and requires {pluginname} to be running in xref:advcode.adoc#advcode_inline[inline mode].

> === Advanced Table 2.0.3
> 
> The {productname} 6.4.1 release includes an accompanying release of the **Advanced Table** premium plugin.
> 
> **Advanced Table** 2.0.3 includes the following addition:
> 
> ==== Updated the **Advanced Table** dropdown menu styles to match the default {productname} styles
> //#TINY-9630
> 
> Previously, the **Advanced Table** premium plugin used browser-native elements for some dropdown menus.
> 
> With this update, these menus are now created using the equivalent {productname} custom element, making these dropdown menus present with the same styling as other {productname} UI elements.
> 
> For information on the **Advanced Table** plugin, see xref:advtable.adoc[Advanced Table].
> 
> 

Reviewer 3: This update is intended for 6.5
```suggestion
```

**Merge Tags** 1.2.1 includes the following change.

==== The mceMergeTag command now preserves the selected content as a plain text and wraps it in a merge tag prefix and suffix if the content length exceeds the limit.
//#TINY-9444

In previous versions of {productname}, the `mceMergeTag` command would not wrap the selected content within the merge tag prefix and suffix when the content length exceeded the character limit for the merge tag.

As a consequence, when the `text_patterns` feature was enabled and a user entered a too-long merge tag, the prefix and suffix of the tag were excluded from the inserted content.

To address this problem, improvements to the `mceMergeTag` command were introduced with the release of {productname} 6.4.1. Now, if the length of the selected content exceeds the character limit for the merge tag, the content will be inserted as a plain text, including the prefix and suffix.

[NOTE]
If any content is embraced with a merge tag prefix and suffix but exceeds the character limit for merge tag content length, it will not be converted into an official merge tag.

For information on the **Merge Tags** plugin, see: xref:mergetags.adoc[Merge Tags].

> 

Reviewer 3: there is another entry of `Merge Tag` in `line 169`

Author: ```suggestion
```

Author: @ArvinJ-H removing duplicate, outdated version.
Good pickup.

> === Page Embed 2.1.1
> 
> The {productname} 6.4.1 release includes an accompanying release of the **Page Embed** premium plugin.
> 
> **Page Embed** 2.1.1 includes the following addition:
> 
> ==== Updated the **Page Embed** dropdown menu styles to match the default {productname} styles
> //#TINY-9630
> 
> Previously, the **Page Embed** premium plugin used browser-native elements for some dropdown menus.
> 
> With this update, these menus are now created using the equivalent {productname} custom element, making these dropdown menus present with the same styling as other {productname} UI elements.
> 
> For information on the **Page Embed** plugin, see xref:pageembed.adoc[Page Embed].
> 
> 
> === Permanent Pen 2.0.3
> 
> The {productname} 6.4.1 release includes an accompanying release of the **Permanent Pen** premium plugin.
> 
> **Permanent Pen** 2.0.3 includes the following addition:
> 
> ==== Updated the **Permanent Pen** dropdown menu styles to match the default {productname} styles
> //#TINY-9630
> 
> Previously, the **Permanent Pen** premium plugin used browser-native elements for some dropdown menus.
> 
> With this update, these menus are now created using the equivalent {productname} custom element, making these dropdown menus present with the same styling as other {productname} UI elements.
> 
> For information on the **Permanent Pen** plugin, see xref:permanentpen.adoc[Permanent Pen].
> 
> 

Reviewer 3: This update is intended for 6.5
```suggestion
```

For information on using premium skins and icon packs, see: xref:premium-skins-and-icons.adoc[Premium Skins and Icon Packs].

[[improvements]]
== Improvements

{productname} 6.4.1 also includes the following improvements:

=== Direct invalid child text nodes of list elements will be wrapped in list item elements.
//#TINY-4818

Invalid child nodes were throwing a `Uncaught TypeError:` when `<ol>` tags directly included a text node. This `TypeError` would not allow the user to `Tab` a text node while nested within `<ol>/<li>` HTML tags.

Improvements were made to add a content filter for lists within {productname} 6.4.1. This improvement meant that any _invalid child nodes_ nested within `<li>` elements would be wrapped with this filter to avoid the `Uncaught TypeError:`.

In {productname} 6.4.1, direct _invalid child text nodes_ of list elements are now wrapped in list item elements, which allow the user to now `Tab` nested text nodes that are within `<ol>/<li>` tags.

=== Templates will be parsed before preview and insertion to make preview consistent with inserted template content and prevent XSS
//#TINY-9244

> In previous versions of {productname}, templates were being parsed before `preview` and `insertion` to make preview consistent with content being inserted into the template to prevent Cross-Site Scripting (XSS) issues.
> As a consequence, the Templates plugin would not automatically format or sanitise the HTML passed to it as templates. This means that when previewing templates, the content may look different compared to when it is actually inserted into the editor. This can be a security issue, as unsanitized HTML can allow for Cross-Site Scripting (XSS), which can be harmful.
> 
> To prevent this from happening, {productname} made improvements to the Template plugin that now sanitises content when retrieving preview content, and again before running `editor.dom.create` on it during the template insertion action. By doing this, all preview content will now match the template when it is inserted into the editor, which removed the possibility of any XSS issues arising.
> 
> In {productname} 6.4.1, Templates are now parsed before preview and insertion to make preview consistent with inserted template content and prevent XSS from occuring.

Reviewer 2: ```suggestion
In previous versions of {productname}, the Template plugin does not parse templates before they are previewed, or before `editor.dom.create` is run on it during insertion. Not parsing the template HTMLs means they are not validated or sanitized, which has two consequences:

. Previewing pre-parsed templates may show a visually different result to after the template is parsed and inserted into the editor; and
. Reading unsanitized HTML when previewing templates and when `editor.dom.create` is run during insertion may allow room for cross-site scripting (XSS), which can be a security issue.

To address this, the Template plugin in {productname} 6.4.1 now parses and serializes template HTMLs before it is previewed and before `editor.dom.create` is run on it during template insertion. By doing this, all template previews will now visually match the post-insertion result, and the possibility of XSS issues arising is minimized.
```

=== The editor.insertContent API would insert contents inside noneditable elements if the selection was inside the element
//#TINY-9462

In previous versions of {productname}, the settings for the `contentEditable` property in the editor root was set to `false`. As a consequence, the user could insert content using `api` commands to edit a `contenteditable="false"` root element.

To fix this bug, in {productname} 6.4.1, a blocker was added to the `editor.insertContent` API and `mceInsertContent` command, which stops `contenteditable="false"` root elements from being edited.

In {productname} 6.4.1, inserting contents inside a `contenteditable="false"` element while the selection is inside another element, is no longer possible.

=== Closing a dialog would scroll down the document in Safari
//#TINY-9148

While using previous versions of {productname}, `clicking` away to dismiss the *Search* and *Replace* dialog within Safari would scroll the page to the bottom of the editor. As a result the user would have to manually scroll back to the original `focus` area each time the steps above were executed.

> To correct this issue, a new Css.set was implemented to keep the `focus` on the DOM container in the initial scroll document position. 

Reviewer 5: ```suggestion
To correct this issue, the active element will be blurred before closing the dialog.
```

Maybe? Safari scrolls down to the sink if the dialog is selected before removing the dialog, so the changes made in the ticket were blurring the active element. Then remove the dialog so it doesn't scroll.

. When the selection is collapsed and a keyboard shortcut, namely Ctrl+Backspace (Windows/Linux), Ctrl+Delete (Windows/Linux), Opt+Delete (macOS), Opt+Fn+Delete (macOS), or ⌘+Delete (macOS), is used to delete all content within a format tag and the post-deletion caret is placed at where the start of the content used to be (e.g. `<p>word1 <span style="text-decoration: underline;">word2</span></p>` with the caret positioned after `2`, and Ctrl+Backspace (Windows/Linux) or Opt+Delete (macOS) is performed to delete `word2`).

The recreated formats are not consistent with the original formats. For example, a `<span style="text-decoration: underline;">` tag will be recreated as a `<u>` tag. This can consequently impact the HTML document's semantics and visuals, causing a bug.

A key override on the Backspace and Delete keys already exists to address the first method. However, the bug persists for the second and third methods.

> In {productname} 6.4, two new key overrides have been added to address the second and third method:

Reviewer 2: A typo I made when I suggested these changes.
```suggestion
In {productname} 6.4, two new key overrides have been added to address the second and third methods:
```

> In {productname} 6.4, two new key overrides have been added to address the second and third methods:

Reviewer 2: ```suggestion
In {productname} 6.4.1, two new key overrides have been added to address the second and third methods:
```

[[accompanying-premium-plugin-end-of-life-announcement]]
== Accompanying Premium Plugin end-of-life announcement

The following premium plugin has been announced as reaching its end-of-life:

=== Real-Time Collaboration

TinyMCE’s xref:rtc-introduction.adoc[Real-time Collaboration (RTC)] plugin will be retired and deactivated on December 31, 2023, and is no longer available for purchase.

[[accompanying-premium-skins-and-icon-packs-changes]]
== Accompanying Premium Skins and Icon Packs changes

The {productname} 6.4.1 release includes an accompanying release of the **Premium Skins and Icon Packs**.

=== Premium Skins and Icon Packs

The **Premium Skins and Icon Packs** release includes the following updates:

The **Premium Skins and Icon Packs** were rebuilt to pull in the changes also incorporated into the default {productname} 6.4.1 skin, Oxide.

For information on using premium skins and icon packs, see: xref:premium-skins-and-icons.adoc[Premium Skins and Icon Packs].

[[improvements]]
== Improvements

{productname} 6.4.1 also includes the following improvements:

=== Direct invalid child text nodes of list elements will be wrapped in list item elements.
//#TINY-4818

> Invalid child nodes were throwing a `Uncaught TypeError:` when `<ol>` tags directly included a text node. This `TypeError` would not allow the user to `Tab` a text node while nested within `<ol>/<li>` HTML tags.
> 
> Improvements were made to add a content filter for lists within {productname} 6.4.1. This improvement meant that any _invalid child nodes_ nested within `<li>` elements would be wrapped with this filter to avoid the `Uncaught TypeError:`.
> 
> In {productname} 6.4.1, direct _invalid child text nodes_ of list elements are now wrapped in list item elements, which allow the user to now `Tab` nested text nodes that are within `<ol>/<li>` tags.

Author: ```suggestion
In previous versions of {productname}, when pressing `Tab` while the cursor is placed next to a `<li>` element, a console error would be un expectantly be thrown.

Improvements were made that added a content filter for lists within {productname} 6.4. This improvement meant that any _invalid child nodes_ nested within `<li>` elements would be wrapped with this filter to avoid the `Uncaught TypeError:`.

In {productname} 6.4, pressing `Tab` while the caret is nested within `<li>` tags, no longer throws a error to the console.
```

Author: @TheSpyder Please see suggested changes

Reviewer: Sorry, but no. The bug was not the cursor position alone, it was the cursor position _when the document contained unexpected HTML_. That part is critical to keep.

> a console error would be un expectantly be thrown.

typo here

I also think there are too many commas but that's not a topic for holding up the merge

Reviewer 2: ```suggestion
In previous versions of {productname}, pressing `Tab` while the caret is in a `<ul>` or `<ol>` list containing direct child text nodes would unexpectedly cause an `Uncaught TypeError` console error to be thrown.

In the following example,

[source,html]
----
<ul>
  item1
  <li>item2</li>
</ul>
----

if the caret is placed anywhere on either `item1` or `item2`, pressing `Tab` would throw an `Uncaught TypeError` to the console.

In {productname} 6.4.1, a content filter for `<ul>` and `<ol>` lists was added to wrap any _direct child text nodes_ within an `<li>`. This transforms them into standard `<li>` child elements, preventing the `Uncaught TypeError` from being thrown.
```

Reviewer 2: @kemister85 Your change still contains a few inaccuracies regarding the scopes of the bug and fix. Please use my suggestion.

Author: I committed your suggesting this morning? Let me check

Reviewer 2: Apologies, Github had shown me the wrong view.

> In {productname} 6.4.1, direct _invalid child text nodes_ of list elements are now wrapped in list item elements, which allow the user to now `Tab` nested text nodes that are within `<ol>/<li>` tags.

Reviewer 2: ```suggestion
In previous versions of {productname}, pressing `Tab` while the caret is in a `<ul>` or `<ol>` list containing direct child text nodes would unexpectedly cause an `Uncaught TypeError` console error to be thrown.

In the following example,

[source,html]
----
<ul>
  item1
  <li>item2</li>
</ul>
----

if the caret is placed anywhere on either `item1` or `item2`, pressing `Tab` would throw an `Uncaught TypeError` to the console.

In {productname} 6.4.1, a content filter for `<ul>` and `<ol>` lists was added to wrap any _direct child text nodes_ within an `<li>`. This transforms them into standard `<li>` child elements, preventing the `Uncaught TypeError` from being thrown.
```

In {productname} 6.4.1, pressing backspace in an empty line now preserves the inline formats of the previous empty line.

=== Pressing enter inside the `fontsizeinput` input would not move the focus back into the editor content
//#TINY-9598

During development for the new `fontsizeinput` toolbar item, an issue was found that by pressing enter inside the input field would not move the `focus` back into the {productname} editor body. 

As a consequence, any changes made to the font size within the new toolbar input field, would result in the user having no way to use the new font size within the editor.

After some improvements to the `fontsizeinput` toolbar item, pressing enter in the input field now sets the focus back to the {productname} editor viewport, which allows the new font size to be used. 

In {productname} 6.4.1, pressing enter inside the `fontsizeinput` input now returns the focus back to the editor content body.

[[additions]]
== Additions

{productname} 6.4.1 also includes the following additions:

> === New tree component that can be used in dialog body panel
> //#TINY-9532
> 
> A new tree component has been created to support the *Advanced Template* premium plugin that works in conjunction with a new tree dialog body panel.
> 
> This tree component is limited to *one* level and is used in the dialog body panel to allow users to organize templates into levels for better UI. The tree component is designed to create a visual representation of the hierarchical structure for the template system, which allows users to easily identify the various templates and levels within each catagory, giving users the ability to quickly and easily navigate through the templates, and select the right one for their needs.
> 
> This new tree component is an important addition to the new Premium Plugin: *Advanced Templates*, which is now available as a premium plugin with editors running {productname} 6.4.1 or later.

Author: ```suggestion
=== New tree component that can be used in dialog body panel
//#TINY-9532

A new tree component has been added to {productname}, that can be used in custom plugins and models.

The new tree component is designed to create a visual representation of the hierarchical structure of items such as, showcasing folders and the files within these folders, similar to how a directory works which allows users to easily identify various structure levels within set categories, such as how the new `advtemplate` works.

This new tree component is an important addition to {productname} which will be used in several features with up and coming releases in the future.
```

Author: @hamza0867 I found the suggested changes, they were outdated and not visible.

Reviewer 10: This one sounds good to me

//#TINY-9462

In previous versions of {productname}, elements within a `contenteditable="false"` root could be unexpectedly modified by the `editor.insertContent` API and the `mceInsertContent` command..

As a consequence all elements within the root `body` set to `contenteditable="false";` value, were affected by the issue.

The new additions and improvements with the introduction of {productname} 6.4.1, include a new `isEditable` API been added to the `editor.dom` which now blocks the `editor.insertContent` API and `mceInsertContent` command when the contents return a `true` value if the editor detects the element is a `contenteditable="false";` in a `contenteditable="false";` root.

As a result, any elements that return a `true` value when a specified element is `contenteditable="true"`, will call the new `isEditable` API to action the new blocker to ensure `contenteditable="false";` elements cannot be modified when they are within a root that is set to `contenteditable="false"`.

=== New setText and setIcon methods added to menu button and toolbar button APIs
//#TINY-9268

> In previous versions of {productname}, it was not possible to change the `text` or `icon` of a toolbar button after it had been set for the first time. As a consequence, the menu and toolbar button API, would not extend or update the parent `icon` or `text` directly, based upon what the user had selected within the menubar or toolbar.
> 
> In {productname} 6.4.1, two new additions were added to the menu button API.
> 
> . Added `setText` and `setIcon` properties to both the toolbar button `api`, and menu button `api`.
> . Provided the menu button `api` property as an *argument* which will fetch a function that will allow menu items to change the `icon` or `text` of the parent menu button.
> 
> As a result, in {productname} 6.4.1, menu items can change based on the parent menu button `api` properties.

Author: ```suggestion
In {productname} 6.4, two new methods (`setText` and `setIcon` ) were added to the menu and toolbar button APIs, that now update the parent icon or text based on what menu is selected from the button.

As a result, in {productname} 6.4, menu items can now change based on the parent menu and toolbar button selection.
```

Author: @tiny-ben-tran - Moved this suggestion here due to being outdated, can you please review when possible.

Reviewer 6: ```suggestion
In previous versions of {productname}, it was not possible to change the `text` or `icon` of a toolbar or menu button dynamically.

 In {productname} 6.4, two new methods `setText` and `setIcon` were added to the menu and toolbar button APIs, to enable updating the text and icon of the buttons after the editor initialization.
```

Author: ```suggestion
In previous versions of {productname}, it was not possible to change the `text` or `icon` of a toolbar or menu button dynamically.

In {productname} 6.4, two new methods `setText` and `setIcon` were added to the menu and toolbar button APIs, to enable updating the `text` and `icon` of the buttons after the editor initialization.
```

=== New fontsizeinput toolbar item which allows the user to set the size via input and also increase and decrease it with + and - buttons
//#TINY-9429

In previous versions of {productname}, users did not have the ability to set their font size by either direct input into a component field or by using plus `+` or minus `-` buttons.

As a consequence, {productname} introduced a new UI component that gives users an option to set the default unit of measure (UOM) for font sizes via the new input field or the plus `+` or minus `-` buttons within the toolbar when the `fontsizeinput` is set within the tinymce.init() such as the below:

[source,js]
----
tinymce.init({
    selector: "textarea",  // change this value according to your HTML
    toolbar: "fontsizeinput",
    font_size_input_default_unit: "pt"
});
----

[NOTE]
More details about this component and the possible values that can be used within the new `fontsizeinput` input field can be found in `xref:available-toolbar-buttons.adoc#the-core-toolbar-buttons[fontsizeinput]`.

=== Added skipFocus option to the ToggleToolbarDrawer command to preserve focus
//#TINY-9337

> A focusing event that was in the initial *FloatingToolbarButton* config, in previous versions of {productname} in (2019) was causing `focus` issues with the *ToggleToolbarDrawer*. As a consequence, this meant that while the editor was been scrolled into `focus`(viewport position), and the *ToggleToolbarDrawer* command `skipFocus: boolean` was set, the toolbar draw could still take `focus` when It was not meant to.

Author: ```suggestion
A focusing event that was in the initial *FloatingToolbarButton* config, in previous versions of {productname} in (2019) was causing `focus` issues with the *ToggleToolbarDrawer*. As a consequence, this meant that while the editor was scrolled into `focus`(viewport position), and the *ToggleToolbarDrawer* command `skipFocus: boolean` was set, the toolbar draw could still take `focus` when it was not meant to.
```

Author: @tiny-ben-tran moved your suggestion copy changes to here so i can merge.

Notes:
============
[tiny-ben-tran]
Also, for and the *ToggleToolbarDrawer* command skipFocus: boolean was set, the toolbar draw could still take focus when it was not meant to., does it matter whether skipFocus set to false or true?

@[kemister85]
@tiny-ben-tran does not seem to have any affect if the skipFocus set to false or true as it always returns true

In previous versions of {productname}, pasting links to `text` selections in Safari would sometimes not generate the correct undo stack.

This is due to Safari not supporting the ability to track changes to a page within the editor, which can create issues for users when pasting links to `text` selections, as the undo stack would not be able to recognise the link as a separate element, resulting in the undo action when called, to not remove the hyperlink.

To fix this issue, a workaround was implemented in {productname} 6.4.1, that now bypasses the default paste functionality for this undo function.

In {productname} 6.4.1, pasting links to text selections now generates the correct undo stack on Safari.

=== Toolbar split buttons in *Advanced List* plugin to show the correct state when the cursor is in a checklist
//#TINY-5167

In previous versions of {productname}, when a cursor was placed within a checklist item, the icon for both the checklist and the bullet list would show as `active`. This created UX issues, as having both buttons showing an `active` state was confusing for the user, specifically for navigation purposes when using checklists from the *Advanced List* plugin.

> To fix this issue, {productname} added improvements to the list configuration to checks a new set of conditionals which would determine if the `node` selected matched the parent `node` within the checklist. If the match returns as `true`, the (api) will set only the checklist icon to a `active` state.
> 
> As a result, the checklist `icon` in {productname} 6.4.1, related to the `advlist` plugin will only appear in a `active` state, when the `node` checklist meets the new conditions implemented.

Author: ```suggestion
{productname} addressed the issue by enhancing the list configuration to include a new set of conditionals. These conditionals check if the selected node matches the parent node within the checklist. If there is a match, the (api) will activate only the checklist icon.

Consequently, the checklist icon of {productname} 6.4, which is associated with the **Advanced List** plugin, will only be displayed in an active state if the node checklist fulfills the newly implemented conditions.
```

Author: @tiny-ben-tran Added improvements to this suggestion, if you can review please.

=== Selection of images, hr, tables or noneditable elements was possible if they were within a noneditable element
//#TINY-9473

In previous versions of {productname}, `contenteditable="false"` elements that were nested within another such as a element set with the `contenteditable="false"` attribute, would avoid the `contenteditable="false"` guidelines set. As a consequence, selection of images, hrs, tables or `contenteditable="false"` elements was possible if they were nested within a `contenteditable="false"` element.

Improvements were made to check if the DOM parent node is a `contenteditable="false"` element.

In {productname} 6.4.1, elements with `contenteditable="false"` attributes can no longer be selected when they are children of another `contenteditable="false"` attribute, such as:

.Example of `contenteditable="false"` elements inside `contenteditable="false"` root.
[source,HTML]
----
<div> Parent div element
  <div contenteditable="false">First child noneditable div element</div>
  <div contenteditable="true">Second child editable div element</div>
</div>
----

> === Ranged deletion of formatted text using selection or keyboard shortcut would sometimes cause Blink and Webkit browsers to insert interpreted tags upon typing, which may result in inconsistent tags

Author: ```suggestion
=== Ranged deletion of formatted text using selection or keyboard shortcut would sometimes cause Blink and Webkit browsers to recreate tags upon typing, which may result in inconsistent tags
```

//#TINY-9302

> Blink & Webkit browsers (Chrome & Safari) were deleting the contents of inline formats which would cause the format to be lost and not retained. As a consequence, this issue would impact the semantics and visuals for the *HTML* document, which is not be visible to the user.
> 
> A override fix already exists for _character-by-character_ deletion using backspace/delete keys with a collapsed selection. However, this particular bug persists when the format is deleted when there is a ranged selection that starts at the beginning of the inline format(s) and encompasses at least the entirety of the inline format(s). The bug also occurs when deletion is done using common keyboard shortcuts including `Ctrl/Alt + Backspace`, `Ctrl/Alt + Delete`, and `Cmd + Backspace`.
> 
> In {productname} 6.4.1, two new key overrides have been added to:
> 
> . Check that a `non-collapsed` selection satisfies the conditions set (specifically for a ranged selection using `backspace/delete`).
> . Check a range of conditions *before* continuing (for keyboard shortcut `backspace/delete`).
> 
> As a result, the ranged selection for backspace/delete now check that all of the original formats at the cursor selection *start*, before deletion, are retained with either the native browser behaviour or a caret. For keyboard shortcuts that are using backspace/delete, will create a new unformatted caret which prevents browsers from inserting interpolated formats.

Author: ```suggestion
The native `contenteditable` behaviour on Blink & Webkit browsers (Chrome & Safari) does not retain the original format(s) at the caret after all content within an inline format tag has been deleted. Upon typing again, the browser will attempt to recreate the formats, if one of the following methods is used to delete all content in the format tag:

. When the selection is collapsed and the last character or element of the content of a format tag is deleted by pressing Backspace or Delete (e.g. `<p><span style="text-decoration: underline;">a</span></p>` with the caret positioned after `a`);
. When a ranged selection has been made from the start of the content of a format tag to at or after the end of the content (to include the content's entirety), and the ranged selection is deleted by pressing Backspace or Delete (e.g. `<p><span style="text-decoration: underline;">ab</span>c</p>` with the selection made to include `ab`); or
. When the selection is collapsed and a keyboard shortcut, namely Ctrl+Backspace (Windows/Linux), Ctrl+Delete (Windows/Linux), Opt+Delete (macOS), Opt+Fn+Delete (macOS), or ⌘+Delete (macOS), is used to delete all content within a format tag and the post-deletion caret is placed at where the start of the content used to be (e.g. `<p>word1 <span style="text-decoration: underline;">word2</span></p>` with the caret positioned after `2`, and Ctrl+Backspace (Windows/Linux) or Opt+Delete (macOS) is performed to delete `word2`).

The recreated formats are not consistent with the original formats. For example, a `<span style="text-decoration: underline;">` tag will be recreated as a `<u>` tag. This can consequently impact the HTML document's semantics and visuals, causing a bug.

A key override on the Backspace and Delete keys already exists to address the first method. However, the bug persists for the second and third methods.

In {productname} 6.4, two new key overrides have been added to address the second and third method:

* For Method 2: When Backspace or Delete is detected on `keydown` on a ranged selection, the override checks if the conditions for Method 2 is satisfied. If so, the override will read the current formats, delete the selection, and then create a caret containing the original formats.
* For Method 3: When one of the keyboard shortcuts from Method 3 is detected on `keyup`, the override creates an unformatted caret at the current selection. Performing the override on `keyup` allows the native behaviour of the deletion shortcuts to occur. Then, applying an unformatted caret at the selection after deletion will prevent the browser from recreating any formats.
```

=== Visual characters were rendered inside noneditable elements
//#TINY-9474

In previous versions of {productname}, a bug was found that affected the `visual characters` plugin. As a consequence, the visual characters selector would render inside `contenteditable="false"` elements after the plugin was applied to the editor.

> To fix this issue, a conditional filter was added which identifies, if an element is either a `contenteditable="true"` element or a `contenteditable="false"` element to its decendants.  If the element matches as a `contenteditable="false"` element to itself or decendants, the `visual character` plugin will remove the highlighted selecter from within the `dom` viewport.

Author: ```suggestion
In order to address this problem, improvements were made that now detects whether an element or any of its descendants has the `contenteditable="true"` or `contenteditable="false"` attribute. If the element is identified as a `contenteditable="false"` element, either by itself or through its descendants, the visual character plugin will eliminate the highlighted selector from the dom viewport.
```

Author: @tiny-ben-tran reposting this as the old suggestion was outdated.

Reviewer 6: ```suggestion
In order to address this problem, improvements were made to better detect noneditable context, when an element is identified as a `contenteditable="false"` element, either by itself or through its descendants.

```

As a result, in {productname} 6.4.1, *list* or *Advanced List*  plugins that are set as `contenteditable = "false"` elements can no longer be modified by using list API commands, toolbar buttons and menu items.

=== Color picker dialog would not update the preview color if the hex input value was prefixed with # symbol
//#TINY-9457

In previous versions of {productname}, the preview sample box within the color picker dialog would not render. This issue could be replicated when a hex `value` was prefixed with a `#` symbol.

Example of a `incorrect` vs `correct` value input:

* `#BDE123`: Incorrect hex value input.
* `BDE123`: Correct hex value input.

In order to fix this bug, {productname} 6.4.1, not detects if the user inserts a `#` symbol before the hex `value`, which removes the leading `#` prefix before rendering to the color picker component.

In {productname} 6.4.1, the color picker works as expected when using hex `values` with the `#` prefix.

=== Table cell selection was possible even if the element was in a noneditable root element
//#TINY-9459

> Previous versions of {productname}, table cells set with the `contenteditable="false"` element properties could still be modified. This was because the `contenteditable="true"` element property settings were set to `false` in the root. This meant that any table UI actions, commands or table selections could still affect any tables that were within that specific `contenteditable="false"` element root.
> 
> {productname} 6.4.1, introduced a blocker that is executed when tables are inside a `contenteditable="false"` element root.
> 
> In {productname} 6.4.1, all table UI actions, commands and selections have been `disabled` for tables when either the table or the editor body have been set as `contenteditable="false"` elements.

Author: ```suggestion
In previous versions of {productname}, table cells set within a `contenteditable="false"` root element could still be modified. This was due to configuration settings within the `contenteditable="true"` function as it was set to `false` by default in the root. This meant that any table UI actions, commands or selections could affect tables that were nested within the `contenteditable="false"` root element.
```

Author: @tiny-ben-tran - moved this suggestion here, as the previous suggestion was outdated.
Please review when possible.

Reviewer 6: ```suggestion
In previous versions of {productname}, table cells set within a `contenteditable="false"` root element could still be selected and modified. This meant that any table UI actions, commands or selections could affect tables that were nested within the `contenteditable="false"` root element.

In {productname} 6.4.1, table is no longer selectable or editable by table UI and commands, when either the table or the editor body have been set as `contenteditable="false"` elements.
```

//#TINY-9652

In previous versions of {productname}, when the editor is within a scrollable container, the toolbar would remain visible when the editor is scrolled out of view.

With the release of {productname} 6.4.1, the new `ui_mode` option provides improvements to editors within scrollable contexts, including hiding the toolbar when the editor is out of view when the `ui_mode` option is set to 'split'.

=== `color_cols`: 0 Memory Leak
//#TINY-9649

In previous versions of {productname}, the color swatch would attempt to display 0 columns of colors when the relevant `color_cols`, `color_cols_foreground` or `color_cols_background` options were set to 0. As a consequence, the tab would crash when opening that color swatch.

In  {productname} 6.4.1, the default number of columns, 5, are now used when any of these options is set to 0 or less, preventing this issue from occurring.

[[known-issues]]
== Known issues

This section describes issues that users of {productname} 6.4.1 may encounter and possible workarounds for these issues.

There are several known issues in {productname} 6.4.1.

=== Accessibility check dropdown not populated when switching between issues
//#TINY-9679

> In previous versions of {productname}, a bug was discovered that encountered an issue with the **Accessibility Checker** dropdown. When switching between different issues, the dropdown would not re-populate on existing viewed issues.
> 
> As a result, improvements were made to the **Accessibility Checker** dropdown with {productname} 6.4.1, to ensure that users can now switch between issues with confidence that the dropdown will populate as expected.

Author: ```suggestion
In previous versions of {productname}, a bug was discovered that encountered an issue with the **Accessibility Checker** dropdown. When switching between different issues, the dropdown would not re-populate on existing viewed issues.
```

Author: @MitchCTiny moved this into this review, as it was outdated.
Reference comment: This isn't merged yet, and won't be in 6.4, so it shouldn't be documented as complete. It's in the known issues section.

This new tree component is an important addition to {productname} which will be used in several features with up and coming releases in the future.

=== renderUI property in the Theme type can now return a Promise <RenderResult> instead of RenderResult
//#TINY-9556

In earlier versions of {productname} the `renderUI` function which initiates theme rendering did not wait for the CSS to load. This could cause race conditions with other parts of the editor or customer integrations. As a result {productname} 6.4.1 now waits for a CSS load event before returning from the `renderUI` function. This process necessitates changing the return type of the `renderUI` function; it now returns a `Promise<RenderResult>`. The type is `Promise<RenderResult> | RenderResult` for compatibility, but the `silver` theme always returns a promise.

This change should have no impact on customer integrations, even those implementing custom themes.

=== New isEditable API to editor.selection that returns true or false if the current selection is editable
//#TINY-9462

> In previous versions of {productname}, elements that where within the `editor.selection` could be modified, if they sat within a HTML document body with the properties set to `contenteditable="false"`.
> 
> As a consequence, any `selection | node`, that was encapsulate within the editor body, could be affected by this such as:
> 
> Expected output example:
> 
> [source, HTML]
> ----
> <p>Try to insert content here</p>
> ----
> 
> After executing an `onclick` event within the current `selection | node` using `<button onclick="tinymce.activeEditor.insertContent('text')">insertContent</button>`, the output would modify the existing `contenteditable="false";` element, which should not be possible.
> 
> [source, HTML]
> ----
> <p>Try to insert text content here</p>
> ----
> 
> The new additions and improvements with the introduction of {productname} 6.4.1, include a new `isEditable` API been added to the `editor.selection` which now blocks the `editor.insertContent` API and `mceInsertContent` command when the contents return a `true` value if the editor detects the element is a `contenteditable="false";` element in a `contenteditable="false";` root.
> 
> As a result, any `editor.selection` elements that return a `true` value when a specified element is `contenteditable="true"`, will call the new `isEditable` API to execute the new blocker.

Author: ```suggestion
In previous versions of {productname}, when the body was set to `contenteditable="false"` content could still be inserted with the `insertContent` API.

Example content:
[source, HTML]
----
<body contenteditable="false"><p>Try to insert content here</p></body>
----

After selecting the space between "insert" and "content", executing `tinymce.activeEditor.insertContent('text')` would incorrectly insert the text which should not be possible.

[source, HTML]
----
<body contenteditable="false"><p>Try to inserttextcontent here</p></body>
----

{productname} 6.4 includes a new `isEditable` method on both the `editor.selection` and `editor.dom` APIs which are used to block the `editor.insertContent` API and `mceInsertContent` command when the selection covers a `contenteditable="false"` element.
```

Author: @TheSpyder moved your copy suggestions here as they were outdated from recent changes.
